// ontimize-web-ngx/ontimize/components/contextmenu/context-menu-item/o-context-menu-item.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, EventEmitter, forwardRef } from '@angular/core';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuItems } from '../o-content-menu.class';
export var DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS = [
    'execute'
];
export var DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM = DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS.concat([
    'icon',
    'data',
    'label',
    'oenabled: enabled',
    'svgIcon: svg-icon'
]);
var OContextMenuItemComponent = (function (_super) {
    __extends(OContextMenuItemComponent, _super);
    function OContextMenuItemComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.execute = new EventEmitter();
        _this.type = OComponentMenuItems.TYPE_ITEM_MENU;
        _this.enabled = true;
        return _this;
    }
    OContextMenuItemComponent.prototype.ngOnInit = function () {
        this.enabled = this.parseInput(this.oenabled, true);
    };
    OContextMenuItemComponent.prototype.onClick = function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.triggerExecute(this.data, event);
    };
    OContextMenuItemComponent.prototype.triggerExecute = function (data, $event) {
        if (!this.enabled) {
            return;
        }
        this.execute.emit({ event: $event, data: data });
    };
    Object.defineProperty(OContextMenuItemComponent.prototype, "disabled", {
        get: function () {
            if (this.enabled instanceof Function) {
                return !this.enabled(this.data);
            }
            return !this.enabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OContextMenuItemComponent.prototype, "isVisible", {
        get: function () {
            if (this.ovisible instanceof Function) {
                return this.ovisible(this.data);
            }
            return this.ovisible;
        },
        enumerable: true,
        configurable: true
    });
    OContextMenuItemComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-context-menu-item',
                    template: ' ',
                    inputs: DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM,
                    outputs: DEFAULT_CONTEXT_MENU_ITEM_OUTPUTS,
                    providers: [{ provide: OComponentMenuItems, useExisting: forwardRef(function () { return OContextMenuItemComponent; }) }]
                },] },
    ];
    return OContextMenuItemComponent;
}(OComponentMenuItems));
export { OContextMenuItemComponent };
