// ontimize-web-ngx/ontimize/components/container/row/o-row.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { CommonModule } from '@angular/common';
import { Component, ElementRef, forwardRef, Inject, Injector, NgModule, Optional, ViewEncapsulation } from '@angular/core';
import { MAT_FORM_FIELD_DEFAULT_OPTIONS } from '@angular/material';
import { OSharedModule } from '../../../shared';
import { OFormComponent } from '../../form/form-components';
import { DEFAULT_INPUTS_O_CONTAINER, OContainerComponent } from '../o-container-component.class';
export var DEFAULT_INPUTS_O_ROW = DEFAULT_INPUTS_O_CONTAINER.slice();
var ORowComponent = (function (_super) {
    __extends(ORowComponent, _super);
    function ORowComponent(form, elRef, injector, matFormDefaultOption) {
        var _this = _super.call(this, form, elRef, injector, matFormDefaultOption) || this;
        _this.form = form;
        _this.elRef = elRef;
        _this.injector = injector;
        _this.matFormDefaultOption = matFormDefaultOption;
        return _this;
    }
    ORowComponent.DEFAULT_INPUTS_O_ROW = DEFAULT_INPUTS_O_ROW;
    ORowComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-row',
                    template: "\n    <div #container fxLayout=\"column\" class=\"o-container\" fxFill>\n      <div #containerTitle *ngIf=\"hasHeader()\" fxLayoutAlign=\"start center\" class=\"o-container-title\" layout-padding>\n        <mat-icon *ngIf=\"icon\">{{ icon }}</mat-icon>\n        <span *ngIf=\"title\">{{ title | oTranslate }}</span>\n      </div>\n      <div [class.o-container-gap]=\"hasHeader() || (elevation > 0 && elevation <= 12)\" class=\"o-container-scroll o-scroll\">\n        <div class=\"o-container-outline\" *ngIf=\"isAppearanceOutline() && hasHeader()\">\n          <div class=\"o-container-outline-start\"></div>\n          <div class=\"o-container-outline-gap\"></div>\n          <div class=\"o-container-outline-end\"></div>\n        </div>\n        <div fxLayout=\"row\" fxLayoutAlign=\"{{ layoutAlign }}\" fxLayoutGap=\"{{ layoutGap }}\" fxFlex=\"grow\">\n          <ng-content></ng-content>\n        </div>\n      </div>\n    </div>\n  ",
                    styles: ["\n    .o-row>.o-container{width:100%}.o-row>.o-container .o-container-scroll{overflow:auto;position:relative}\n  "],
                    inputs: DEFAULT_INPUTS_O_ROW,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-row]': 'true',
                        '[class.o-appearance-outline]': 'isAppearanceOutline()',
                        '[class.o-appearance-outline-title]': 'hasTitleInAppearanceOutline()'
                    }
                },] },
    ];
    ORowComponent.ctorParameters = function () { return [
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: ElementRef },
        { type: Injector },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [MAT_FORM_FIELD_DEFAULT_OPTIONS,] }] }
    ]; };
    return ORowComponent;
}(OContainerComponent));
export { ORowComponent };
var ORowModule = (function () {
    function ORowModule() {
    }
    ORowModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ORowComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [ORowComponent]
                },] },
    ];
    return ORowModule;
}());
export { ORowModule };
