// ontimize-web-ngx/ontimize/components/contextmenu/context-menu-separator/o-context-menu-separator.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, forwardRef } from '@angular/core';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS, OComponentMenuItems } from '../o-content-menu.class';
import { OContextMenuItemComponent } from '../o-context-menu-components';
export var DEFAULT_CONTEXT_MENU_ITEM_INPUTS = DEFAULT_INPUTS_O_CONTEXT_MENU_ITEMS.slice();
var OContextMenuSeparatorComponent = (function (_super) {
    __extends(OContextMenuSeparatorComponent, _super);
    function OContextMenuSeparatorComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = OComponentMenuItems.TYPE_SEPARATOR_MENU;
        return _this;
    }
    OContextMenuSeparatorComponent.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
    };
    OContextMenuSeparatorComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-context-menu-separator',
                    template: ' ',
                    inputs: DEFAULT_CONTEXT_MENU_ITEM_INPUTS,
                    providers: [{ provide: OComponentMenuItems, useExisting: forwardRef(function () { return OContextMenuSeparatorComponent; }) }]
                },] },
    ];
    return OContextMenuSeparatorComponent;
}(OContextMenuItemComponent));
export { OContextMenuSeparatorComponent };
