// ontimize-web-ngx/ontimize/services/auth-guard.service.js
import { Injector, Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { LoginService, OUserInfoService } from '../services';
import { Codes } from '../utils';
import { PermissionsService } from './permissions/permissions.service';
var AuthGuardService = (function () {
    function AuthGuardService(injector) {
        this.injector = injector;
        this.router = this.injector.get(Router);
        this.loginService = this.injector.get(LoginService);
        this.oUserInfoService = this.injector.get(OUserInfoService);
        this.permissionsService = this.injector.get(PermissionsService);
    }
    AuthGuardService.prototype.canActivate = function (next, state) {
        var isLoggedIn = this.loginService.isLoggedIn();
        var result = isLoggedIn;
        if (!isLoggedIn) {
            this.permissionsService.restart();
            this.router.navigate([Codes.LOGIN_ROUTE]);
        }
        if (isLoggedIn) {
            this.setUserInformation();
            if (!this.permissionsService.hasPermissions()) {
                result = this.permissionsService.getUserPermissionsAsPromise();
            }
        }
        return result;
    };
    AuthGuardService.prototype.setUserInformation = function () {
        var sessionInfo = this.loginService.getSessionInfo();
        this.oUserInfoService.setUserInfo({
            username: sessionInfo.user,
            avatar: './assets/images/user_profile.png'
        });
    };
    AuthGuardService.decorators = [
        { type: Injectable },
    ];
    AuthGuardService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return AuthGuardService;
}());
export { AuthGuardService };
