// ontimize-web-ngx/ontimize/services/o-modules-info.service.js
import { Injectable, Injector } from '@angular/core';
import { Subject } from 'rxjs';
import { Router, NavigationEnd, ActivatedRoute } from '@angular/router';
import { OTranslateService } from '../services';
var OModulesInfoService = (function () {
    function OModulesInfoService(injector) {
        var _this = this;
        this.injector = injector;
        this.subject = new Subject();
        this.router = this.injector.get(Router);
        this.actRoute = this.injector.get(ActivatedRoute);
        this.translateService = this.injector.get(OTranslateService);
        this.router.events.subscribe(function (ev) {
            if (ev instanceof NavigationEnd) {
                var translation = _this.translateService.get(ev.url);
                if (translation !== ev.url) {
                    _this.setModuleInfo({
                        name: translation
                    });
                }
            }
        });
    }
    OModulesInfoService.prototype.setModuleInfo = function (info) {
        this.storedInfo = info;
        this.subject.next(info);
    };
    OModulesInfoService.prototype.getModuleInfo = function () {
        return this.storedInfo;
    };
    OModulesInfoService.prototype.getModuleChangeObservable = function () {
        return this.subject.asObservable();
    };
    OModulesInfoService.decorators = [
        { type: Injectable },
    ];
    OModulesInfoService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OModulesInfoService;
}());
export { OModulesInfoService };
