// ontimize-web-ngx/ontimize/services/share-can-activate-child.service.js
import { Injector, Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { PermissionsService } from './permissions/permissions.service';
var ShareCanActivateChildService = (function () {
    function ShareCanActivateChildService(injector) {
        this.injector = injector;
        this.router = this.injector.get(Router);
        this.permissionsService = this.injector.get(PermissionsService);
    }
    ShareCanActivateChildService.prototype.setPermissionsGuard = function (guard) {
        this.permissionsGuard = guard;
    };
    ShareCanActivateChildService.prototype.canActivateChildUsingPermissions = function (childRoute, state) {
        if (this.permissionsGuard) {
            return this.permissionsGuard.canActivateChild(childRoute, state);
        }
        return true;
    };
    ShareCanActivateChildService.decorators = [
        { type: Injectable },
    ];
    ShareCanActivateChildService.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return ShareCanActivateChildService;
}());
export { ShareCanActivateChildService };
