// ontimize-web-ngx/ontimize/components/user-info/o-user-info.component.js
import { Component, ElementRef, Injector, NgModule, ViewEncapsulation } from '@angular/core';
import { Router, RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../shared';
import { DialogService, LoginService, OUserInfoService } from '../../services';
export var DEFAULT_INPUTS_O_USER_INFO = [];
export var DEFAULT_OUTPUTS_O_USER_INFO = [];
var OUserInfoComponent = (function () {
    function OUserInfoComponent(elRef, injector, router) {
        var _this = this;
        this.elRef = elRef;
        this.injector = injector;
        this.router = router;
        this.dialogService = this.injector.get(DialogService);
        this.loginService = this.injector.get(LoginService);
        this.oUserInfoService = this.injector.get(OUserInfoService);
        this.userInfo = this.oUserInfoService.getUserInfo();
        this.userInfoSubscription = this.oUserInfoService.getUserInfoObservable().subscribe(function (res) {
            _this.userInfo = res;
        });
    }
    OUserInfoComponent.prototype.ngOnDestroy = function () {
        this.userInfoSubscription.unsubscribe();
    };
    OUserInfoComponent.prototype.onLogoutClick = function () {
        this.loginService.logoutWithConfirmationAndRedirect();
    };
    OUserInfoComponent.prototype.onSettingsClick = function () {
        this.router.navigate(['main/settings']);
    };
    Object.defineProperty(OUserInfoComponent.prototype, "existsUserInfo", {
        get: function () {
            return this.userInfo !== undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OUserInfoComponent.prototype, "avatar", {
        get: function () {
            return this.userInfo ? this.userInfo.avatar : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OUserInfoComponent.prototype, "username", {
        get: function () {
            return this.userInfo ? this.userInfo.username : undefined;
        },
        enumerable: true,
        configurable: true
    });
    OUserInfoComponent.DEFAULT_INPUTS_O_USER_INFO = DEFAULT_INPUTS_O_USER_INFO;
    OUserInfoComponent.DEFAULT_OUTPUTS_O_USER_INFO = DEFAULT_OUTPUTS_O_USER_INFO;
    OUserInfoComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-user-info',
                    inputs: DEFAULT_INPUTS_O_USER_INFO,
                    outputs: DEFAULT_OUTPUTS_O_USER_INFO,
                    template: "\n    <div class=\"o-user-info-container\" fxLayout=\"row\" fxLayoutAlign=\"center\">\n      <ng-container *ngIf=\"existsUserInfo\">\n        <div class=\"o-user-info-wrapper\" fxLayout=\"row\" fxLayoutAlign=\"center center\">\n          <div fxLayout=\"row\" fxLayoutAlign=\"center center\" class=\"o-user-info-avatar-wrapper\">\n            <img [src]=\"avatar\" *ngIf=\"avatar\" />\n          </div>\n          <span class=\"o-user-info-username\" *ngIf=\"username\"> {{ username }}</span>\n          <mat-icon [matMenuTriggerFor]=\"menu\" svgIcon=\"ontimize:keyboard_arrow_down\"></mat-icon>\n        </div>\n        <mat-menu #menu=\"matMenu\" yPosition=\"below\">\n          <button type=\"button\" mat-menu-item (click)=\"onSettingsClick()\">\n            <mat-icon svgIcon=\"ontimize:settings\"></mat-icon>\n            <span>{{ 'SETTINGS' | oTranslate }}</span>\n          </button>\n          <button type=\"button\" mat-menu-item (click)=\"onLogoutClick()\">\n            <mat-icon svgIcon=\"ontimize:power_settings_new\"></mat-icon>\n            <span>{{ 'LOGOUT' | oTranslate }}</span>\n          </button>\n        </mat-menu>\n      </ng-container>\n    </div>\n  ",
                    styles: ["\n    .o-user-info .o-user-info-container{cursor:pointer;height:100%}.o-user-info .o-user-info-avatar-wrapper{width:30px;height:30px;overflow:hidden;border-radius:100%}.o-user-info .o-user-info-avatar-wrapper img{max-width:100%;height:auto}.o-user-info .o-user-info-username{padding-left:6px}.o-user-info mat-icon{padding-left:4px;font-size:16px;line-height:26px}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-user-info]': 'true'
                    }
                },] },
    ];
    OUserInfoComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: Router }
    ]; };
    return OUserInfoComponent;
}());
export { OUserInfoComponent };
var OUserInfoModule = (function () {
    function OUserInfoModule() {
    }
    OUserInfoModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OUserInfoComponent],
                    imports: [CommonModule, OSharedModule, RouterModule],
                    exports: [OUserInfoComponent]
                },] },
    ];
    return OUserInfoModule;
}());
export { OUserInfoModule };
