// ontimize-web-ngx/ontimize/components/bar-menu/menu-separator/o-bar-menu-separator.component.js
import { Component, ViewEncapsulation, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
var OBarMenuSeparatorComponent = (function () {
    function OBarMenuSeparatorComponent() {
    }
    OBarMenuSeparatorComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-bar-menu-separator',
                    template: "\n    <div class=\"fake-li mat-primary separator\"></div>\n  ",
                    styles: ["\n    o-bar-menu-separator .fake-li.separator{border-bottom:2px solid #ccc}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-bar-menu-separator]': 'true'
                    }
                },] },
    ];
    return OBarMenuSeparatorComponent;
}());
export { OBarMenuSeparatorComponent };
var OBarMenuSeparatorModule = (function () {
    function OBarMenuSeparatorModule() {
    }
    OBarMenuSeparatorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OBarMenuSeparatorComponent],
                    imports: [CommonModule],
                    exports: [OBarMenuSeparatorComponent]
                },] },
    ];
    return OBarMenuSeparatorModule;
}());
export { OBarMenuSeparatorModule };
