// ontimize-web-ngx/ontimize/directives/o-tab-group.directive.js
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
var OTabGroupDirective = (function () {
    function OTabGroupDirective(renderer, el) {
        this.renderer = renderer;
        this.el = el;
        this._mode = 'ontimize';
        this._defaultMode = 'ontimize';
    }
    Object.defineProperty(OTabGroupDirective.prototype, "mode", {
        get: function () {
            return this._mode;
        },
        set: function (mode) {
            this._mode = mode;
            this.applyMode();
        },
        enumerable: true,
        configurable: true
    });
    OTabGroupDirective.prototype.applyMode = function (mode) {
        this.renderer.removeClass(this.el.nativeElement, OTabGroupDirective.OTabModes.material);
        this.renderer.removeClass(this.el.nativeElement, OTabGroupDirective.OTabModes.ontimize);
        this.renderer.addClass(this.el.nativeElement, OTabGroupDirective.OTabModes[this.mode || this._defaultMode]);
    };
    OTabGroupDirective.OTabModes = {
        ontimize: 'o-tab-ontimize',
        material: 'o-tab-material'
    };
    OTabGroupDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oTabGroup]'
                },] },
    ];
    OTabGroupDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    OTabGroupDirective.propDecorators = {
        mode: [{ type: Input, args: ['oTabGroup',] }]
    };
    return OTabGroupDirective;
}());
export { OTabGroupDirective };
