// ontimize-web-ngx/ontimize/components/bar-menu/menu-item/o-bar-menu-item.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Inject, Injector, forwardRef, ElementRef, NgModule, ViewEncapsulation } from '@angular/core';
import { RouterModule, Router } from '@angular/router';
import { CommonModule } from '@angular/common';
import { Util } from '../../../utils';
import { OSharedModule } from '../../../shared';
import { OBarMenuComponent } from '../o-bar-menu.component';
import { OBaseMenuItemClass } from '../o-base-menu-item.class';
export var DEFAULT_INPUTS_O_BAR_MENU_ITEM = OBaseMenuItemClass.DEFAULT_INPUTS_O_BASE_MENU_ITEM.concat([
    'route',
    'action'
]);
var OBarMenuItemComponent = (function (_super) {
    __extends(OBarMenuItemComponent, _super);
    function OBarMenuItemComponent(menu, elRef, injector) {
        var _this = _super.call(this, menu, elRef, injector) || this;
        _this.menu = menu;
        _this.elRef = elRef;
        _this.injector = injector;
        _this.router = _this.injector.get(Router);
        return _this;
    }
    OBarMenuItemComponent.prototype.ngOnInit = function () {
        _super.prototype.ngOnInit.call(this);
    };
    OBarMenuItemComponent.prototype.collapseMenu = function (evt) {
        if (this.menu) {
            this.menu.collapseAll();
        }
    };
    OBarMenuItemComponent.prototype.onClick = function () {
        if (this.disabled) {
            return;
        }
        if (Util.isDefined(this.route)) {
            this.router.navigate([this.route]);
        }
        else if (Util.isDefined(this.action)) {
            this.action();
        }
    };
    OBarMenuItemComponent.DEFAULT_INPUTS_O_BAR_MENU_ITEM = DEFAULT_INPUTS_O_BAR_MENU_ITEM;
    OBarMenuItemComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-bar-menu-item',
                    template: "\n    <div class=\"fake-li mat-primary\" *ngIf=\"!restricted\" (click)=\"collapseMenu($event)\" [class.fake-li-hover]=\"isHovered\">\n      <a *ngIf=\"action || route\" class=\"mat-list-item\" (click)=\"onClick()\">\n        <mat-icon class=\"o-bar-menu-item-icon mat-24\">{{ icon }}</mat-icon>\n        <div class=\"o-bar-menu-item-title\">{{ title | oTranslate }}</div>\n      </a>\n      <div *ngIf=\"!route && !action\" class=\"o-bar-menu-item-text mat-list-item\">\n        <mat-icon class=\"o-bar-menu-item-icon mat-24\">{{ icon }}</mat-icon>\n        <div class=\"o-bar-menu-item-title\">{{ title | oTranslate }}</div>\n      </div>\n    </div>\n  ",
                    styles: ["\n    .o-bar-menu-item[disabled=true] .fake-li{opacity:.5}.o-bar-menu-item[disabled=true] .fake-li a.mat-list-item{cursor:default}.o-bar-menu-item .o-bar-menu-item-text{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-box-orient:horizontal;-webkit-box-direction:normal;-webkit-flex-direction:row;-ms-flex-direction:row;flex-direction:row;-webkit-box-align:center;-webkit-align-items:center;-ms-flex-align:center;align-items:center;box-sizing:border-box;font-size:16px;height:48px;padding:0 16px}.o-bar-menu-item .o-bar-menu-item-icon{display:inline-block;margin:6px 10px 0 0}.o-bar-menu-item .o-bar-menu-item-icon ~ .o-bar-menu-item-title{padding-left:0}.o-bar-menu-item .o-bar-menu-item-title{display:inline-block;position:relative;top:3px;padding-left:40px}\n  "],
                    inputs: DEFAULT_INPUTS_O_BAR_MENU_ITEM,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-bar-menu-item]': 'true',
                        '[attr.disabled]': 'disabled'
                    }
                },] },
    ];
    OBarMenuItemComponent.ctorParameters = function () { return [
        { type: OBarMenuComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OBarMenuComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return OBarMenuItemComponent;
}(OBaseMenuItemClass));
export { OBarMenuItemComponent };
var OBarMenuItemModule = (function () {
    function OBarMenuItemModule() {
    }
    OBarMenuItemModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OBarMenuItemComponent],
                    imports: [OSharedModule, CommonModule, RouterModule],
                    exports: [OBarMenuItemComponent]
                },] },
    ];
    return OBarMenuItemModule;
}());
export { OBarMenuItemModule };
