// ontimize-web-ngx/ontimize/components/snackbar/o-snackbar.component.js
import { CommonModule } from '@angular/common';
import { Component, Injector, NgModule } from '@angular/core';
import { MatSnackBarRef } from '@angular/material';
import { OTranslateModule } from '../../pipes/o-translate.pipe';
import { OSharedModule } from '../../shared';
var OSnackBarConfig = (function () {
    function OSnackBarConfig() {
    }
    return OSnackBarConfig;
}());
export { OSnackBarConfig };
var OSnackBarComponent = (function () {
    function OSnackBarComponent(injector) {
        this.injector = injector;
        this.iconPosition = 'left';
        this.snackBarRef = this.injector.get(MatSnackBarRef);
    }
    OSnackBarComponent.prototype.open = function (message, config) {
        this.message = message;
        if (config) {
            if (config.action) {
                this.action = config.action;
            }
            if (config.icon) {
                this.icon = config.icon;
            }
            if (config.iconPosition) {
                this.iconPosition = config.iconPosition;
            }
        }
    };
    OSnackBarComponent.prototype.onAction = function (e) {
        this.snackBarRef.closeWithAction();
    };
    OSnackBarComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-snackbar',
                    template: "\n    <div fxLayout=\"row\" fxLayoutAlign=\"start center\">\n      <mat-icon *ngIf=\"icon && iconPosition==='left'\" class=\"o-snackbar-icon\">{{ icon }}</mat-icon>\n      <span fxFlex class=\"o-snackbar-message\">\n        {{ message | oTranslate }}\n      </span>\n      <mat-icon *ngIf=\"icon && iconPosition==='right'\" class=\"o-snackbar-icon\">{{ icon }}</mat-icon>\n      <button type=\"button\" *ngIf=\"action\" (click)=\"onAction($event)\" class=\"mat-simple-snackbar-action\">{{ action | oTranslate }}</button>\n    </div>\n  ",
                    styles: ["\n    .o-snackbar-message{padding:0 8px}.mat-simple-snackbar-action{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer;outline:none;border:none;-webkit-tap-highlight-color:transparent;background:transparent}\n  "],
                    host: {
                        '[class.o-snackbar]': 'true'
                    }
                },] },
    ];
    OSnackBarComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OSnackBarComponent;
}());
export { OSnackBarComponent };
var OSnackBarModule = (function () {
    function OSnackBarModule() {
    }
    OSnackBarModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OSnackBarComponent],
                    imports: [CommonModule, OSharedModule, OTranslateModule],
                    exports: [CommonModule, OSnackBarComponent]
                },] },
    ];
    return OSnackBarModule;
}());
export { OSnackBarModule };
