// ontimize-web-ngx/ontimize/components/contextmenu/o-context-menu.directive.js
import { Directive, HostListener, Injector } from '@angular/core';
import { OContextMenuService } from './o-context-menu.service';
export var DEFAULT_CONTEXT_MENU_DIRECTIVE_INPUTS = [
    'oContextMenu',
    'oContextMenuData'
];
var OContextMenuDirective = (function () {
    function OContextMenuDirective(injector) {
        this.injector = injector;
        this.oContextMenuService = this.injector.get(OContextMenuService);
    }
    OContextMenuDirective.prototype.onRightClick = function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.oContextMenuService.showContextMenu.next({
            contextMenu: this.oContextMenu,
            event: event,
            data: this.oContextMenuData
        });
    };
    OContextMenuDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oContextMenu]',
                    inputs: DEFAULT_CONTEXT_MENU_DIRECTIVE_INPUTS
                },] },
    ];
    OContextMenuDirective.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OContextMenuDirective.propDecorators = {
        onRightClick: [{ type: HostListener, args: ['contextmenu', ['$event'],] }]
    };
    return OContextMenuDirective;
}());
export { OContextMenuDirective };
