// ontimize-web-ngx/ontimize/shared/material/custom.material.module.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { Inject, Injectable, NgModule, Optional } from '@angular/core';
import { DateAdapter, MAT_DATE_FORMATS, MAT_DATE_LOCALE, MatAutocompleteModule, MatButtonModule, MatButtonToggleModule, MatCardModule, MatCheckboxModule, MatChipsModule, MatDatepickerModule, MatDialogModule, MatExpansionModule, MatFormFieldModule, MatGridListModule, MatIconModule, MatInputModule, MatListModule, MatMenuModule, MatPaginatorModule, MatProgressBarModule, MatProgressSpinnerModule, MatRadioModule, MatRippleModule, MatSelectModule, MatSidenavModule, MatSliderModule, MatSlideToggleModule, MatSnackBarModule, MatSortModule, MatTableModule, MatTabsModule, MatToolbarModule, MatTooltipModule, } from '@angular/material';
import { MatMomentDateModule, MomentDateAdapter } from '@angular/material-moment-adapter';
import moment from 'moment';
import { dateFormatFactory } from '../../services/mat-date-formats.factory';
import { OntimizeMatIconRegistry } from '../../services/ontimize-icon-registry.service';
var MATERIAL_MODULES = [
    MatAutocompleteModule,
    MatButtonModule,
    MatButtonToggleModule,
    MatCardModule,
    MatChipsModule,
    MatCheckboxModule,
    MatDatepickerModule,
    MatDialogModule,
    MatExpansionModule,
    MatGridListModule,
    MatIconModule,
    MatInputModule,
    MatFormFieldModule,
    MatListModule,
    MatMenuModule,
    MatProgressBarModule,
    MatProgressSpinnerModule,
    MatRadioModule,
    MatRippleModule,
    MatSelectModule,
    MatSidenavModule,
    MatSliderModule,
    MatSlideToggleModule,
    MatSnackBarModule,
    MatTabsModule,
    MatToolbarModule,
    MatTooltipModule,
    MatMomentDateModule,
    MatTableModule,
    MatPaginatorModule,
    MatSortModule,
    OverlayModule
];
var OntimizeMomentDateAdapter = (function (_super) {
    __extends(OntimizeMomentDateAdapter, _super);
    function OntimizeMomentDateAdapter(dateLocale) {
        return _super.call(this, dateLocale) || this;
    }
    OntimizeMomentDateAdapter.prototype.format = function (date, displayFormat) {
        return _super.prototype.format.call(this, date, this.oFormat || displayFormat);
    };
    OntimizeMomentDateAdapter.prototype.parse = function (value, parseFormat) {
        return _super.prototype.parse.call(this, value, this.oFormat || parseFormat);
    };
    OntimizeMomentDateAdapter.prototype.deserialize = function (value) {
        var date;
        if (typeof value === 'number') {
            date = moment(value);
        }
        if (date && this.isValid(date)) {
            return date;
        }
        return _super.prototype.deserialize.call(this, value);
    };
    OntimizeMomentDateAdapter.decorators = [
        { type: Injectable },
    ];
    OntimizeMomentDateAdapter.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_LOCALE,] }] }
    ]; };
    return OntimizeMomentDateAdapter;
}(MomentDateAdapter));
export { OntimizeMomentDateAdapter };
var ɵ0 = dateFormatFactory;
var OCustomMaterialModule = (function () {
    function OCustomMaterialModule() {
    }
    OCustomMaterialModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    exports: MATERIAL_MODULES,
                    providers: [{
                            provide: DateAdapter,
                            useClass: OntimizeMomentDateAdapter,
                            deps: [MAT_DATE_LOCALE]
                        }, {
                            provide: MAT_DATE_FORMATS,
                            useFactory: ɵ0
                        }, {
                            provide: OntimizeMatIconRegistry,
                            useClass: OntimizeMatIconRegistry
                        }]
                },] },
    ];
    return OCustomMaterialModule;
}());
export { OCustomMaterialModule };
export { ɵ0 };
