// ontimize-web-ngx/ontimize/components/image/fullscreen/fullscreen-dialog.component.js
import { Component, Inject, ViewEncapsulation } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
import { Util } from '../../../util/util';
var OFullScreenDialogComponent = (function () {
    function OFullScreenDialogComponent(dialogRef, data) {
        this.dialogRef = dialogRef;
        this.data = data;
        if (Util.isDefined(data)) {
            this.imageSrc = data;
        }
    }
    OFullScreenDialogComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-fullscreen-dialog',
                    template: "\n    <mat-toolbar color=\"primary\" class=\"mat-elevation-z5\">\n      <span fxFlex></span>\n      <button type=\"button\" mat-icon-button mat-dialog-close>\n        <mat-icon svgIcon=\"ontimize:close\"></mat-icon>\n      </button>\n    </mat-toolbar>\n    <mat-dialog-content fxLayoutAlign=\"center center\" class=\"o-fullscreen-dialog-content\">\n      <img [src]=\"imageSrc\" />\n    </mat-dialog-content>\n  ",
                    styles: ["\n    .o-image-fullscreen-dialog-cdk-overlay .mat-dialog-container{padding:0}.o-image-fullscreen-dialog-cdk-overlay .mat-dialog-container .o-fullscreen-dialog{height:100%;display:block}.o-image-fullscreen-dialog-cdk-overlay .mat-dialog-container .o-fullscreen-dialog .mat-dialog-content.o-fullscreen-dialog-content{margin:12px;height:100%;max-height:calc(100% - 64px - 24px)}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-fullscreen-dialog]': 'true'
                    }
                },] },
    ];
    OFullScreenDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    return OFullScreenDialogComponent;
}());
export { OFullScreenDialogComponent };
