// ontimize-web-ngx/ontimize/services/permissions/error403/o-error-403.component.js
import { Component, ViewEncapsulation, Injector } from '@angular/core';
import { Router } from '@angular/router';
import { Codes } from '../../../utils';
import { NavigationService } from '../../navigation.service';
var Error403Component = (function () {
    function Error403Component(injector) {
        this.injector = injector;
        this.router = this.injector.get(Router);
        this.navigationService = this.injector.get(NavigationService);
        this.lastPageData = this.navigationService.getLastItem();
    }
    Error403Component.prototype.onNavigateBackClick = function () {
        var extras = {};
        var route = '';
        if (this.lastPageData) {
            extras[Codes.QUERY_PARAMS] = this.lastPageData.queryParams;
            route = this.lastPageData.url;
        }
        this.router.navigate([route], extras);
    };
    Error403Component.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-error-403',
                    template: "\n    <div fxFlex fxFlexFill fxLayout=\"column\" fxLayoutAlign=\"center center\">\n      {{ 'MESSAGES.ERROR_403_TEXT' | oTranslate }}\n      <button color=\"accent\" type=\"button\" mat-raised-button (click)=\"onNavigateBackClick()\">\n        {{ 'MESSAGES.ERROR_403_TEXT_BUTTON' | oTranslate }}\n      </button>\n    </div>\n  ",
                    styles: ["\n    .o-error-403{height:100%}.o-error-403 button{margin:16px}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-error-403]': 'true'
                    }
                },] },
    ];
    Error403Component.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return Error403Component;
}());
export { Error403Component };
