// ontimize-web-ngx/ontimize/components/filter-builder/o-filter-builder-query.directive.js
import { Directive, Optional } from '@angular/core';
import { Util } from '../../utils';
import { OFilterBuilderComponent } from './o-filter-builder-components';
var OFilterBuilderQueryDirective = (function () {
    function OFilterBuilderQueryDirective(filterBuilder) {
        if (Util.isDefined(filterBuilder)) {
            this._filterBuilder = filterBuilder;
        }
    }
    OFilterBuilderQueryDirective.prototype.onClick = function (e) {
        if (this._filterBuilder) {
            this._filterBuilder.triggerReload();
        }
    };
    OFilterBuilderQueryDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oFilterBuilderQuery]',
                    inputs: [
                        '_filterBuilder: oFilterBuilderQuery'
                    ],
                    host: {
                        '(click)': 'onClick($event)'
                    },
                    exportAs: 'oFilterBuilderQuery'
                },] },
    ];
    OFilterBuilderQueryDirective.ctorParameters = function () { return [
        { type: OFilterBuilderComponent, decorators: [{ type: Optional }] }
    ]; };
    return OFilterBuilderQueryDirective;
}());
export { OFilterBuilderQueryDirective };
