// ontimize-web-ngx/ontimize/components/filter-builder/o-filter-builder-clear.directive.js
import { Directive, Optional } from '@angular/core';
import { Util } from '../../utils';
import { OFilterBuilderComponent } from './o-filter-builder-components';
var OFilterBuilderClearDirective = (function () {
    function OFilterBuilderClearDirective(filterBuilder) {
        if (Util.isDefined(filterBuilder)) {
            this._filterBuilder = filterBuilder;
        }
    }
    OFilterBuilderClearDirective.prototype.onClick = function (e) {
        if (this._filterBuilder) {
            this._filterBuilder.clearFilter();
        }
    };
    OFilterBuilderClearDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[oFilterBuilderClear]',
                    inputs: [
                        '_filterBuilder: oFilterBuilderClear'
                    ],
                    host: {
                        '(click)': 'onClick($event)'
                    },
                    exportAs: 'oFilterBuilderClear'
                },] },
    ];
    OFilterBuilderClearDirective.ctorParameters = function () { return [
        { type: OFilterBuilderComponent, decorators: [{ type: Optional }] }
    ]; };
    return OFilterBuilderClearDirective;
}());
export { OFilterBuilderClearDirective };
