// ontimize-web-ngx/ontimize/components/list/renderers/o-list-item-card.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, forwardRef, Inject, Injector, NgModule, Optional, Renderer2, ViewEncapsulation } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OListItemComponent } from '../list-item/o-list-item.component';
import { OListItemCardRenderer } from './o-list-item-card-renderer.class';
export var DEFAULT_INPUTS_O_LIST_ITEM_CARD = OListItemCardRenderer.DEFAULT_INPUTS_O_CARD_RENDERER.slice();
export var DEFAULT_OUTPUTS_O_LIST_ITEM_CARD = OListItemCardRenderer.DEFAULT_OUTPUTS_O_CARD_RENDERER.slice();
var OListItemCardComponent = (function (_super) {
    __extends(OListItemCardComponent, _super);
    function OListItemCardComponent(elRef, _renderer, _injector, _listItem) {
        return _super.call(this, elRef, _renderer, _injector, _listItem) || this;
    }
    OListItemCardComponent.prototype.ngAfterViewInit = function () {
        this.modifyMatListItemElement();
    };
    OListItemCardComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-list-item-card',
                    template: "\n    <mat-card fxLayout=\"column\" fxLayoutAlign=\"center center\">\n      <mat-card-title-group>\n        <img *ngIf=\"showImage\" src=\"{{ image }}\" [class.mat-card-sm-image]=\"compareListHeight('small')\" [class.mat-card-md-image]=\"compareListHeight('medium')\"\n          [class.mat-card-lg-image]=\"compareListHeight('large')\">\n        <mat-card-title *ngIf=\"title !== undefined\"> {{ title }}</mat-card-title>\n        <mat-card-subtitle *ngIf=\"subtitle !== undefined\"> {{ subtitle }}</mat-card-subtitle>\n      </mat-card-title-group>\n      <mat-card-actions>\n        <button type=\"button\" mat-button *ngIf=\"action1Text !== undefined\" (click)=\"onAction1ButtonClick($event)\">{{ action1Text }}</button>\n        <button type=\"button\" mat-button *ngIf=\"action2Text !== undefined\" (click)=\"onAction2ButtonClick($event)\">{{ action2Text }}</button>\n      </mat-card-actions>\n    </mat-card>\n  ",
                    styles: ["\n    .mat-list .mat-list-item.o-card-item,.mat-list .mat-list-item.o-card-item .mat-list-item-content{height:auto}.o-list-item-card{padding:8px 0}.o-list-item-card mat-card,.o-list-item-card mat-card-actions,.o-list-item-card mat-card-title-group{width:100%}\n  "],
                    inputs: DEFAULT_INPUTS_O_LIST_ITEM_CARD,
                    outputs: DEFAULT_OUTPUTS_O_LIST_ITEM_CARD,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-custom-list-item]': 'true',
                        '[class.o-list-item-card]': 'true'
                    }
                },] },
    ];
    OListItemCardComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector },
        { type: OListItemComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OListItemComponent; }),] }] }
    ]; };
    return OListItemCardComponent;
}(OListItemCardRenderer));
export { OListItemCardComponent };
var OListItemCardModule = (function () {
    function OListItemCardModule() {
    }
    OListItemCardModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OListItemCardComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OListItemCardComponent]
                },] },
    ];
    return OListItemCardModule;
}());
export { OListItemCardModule };
