// ontimize-web-ngx/ontimize/components/list/renderers/o-list-item-card-renderer.class.js
import { EventEmitter } from '@angular/core';
import { Util } from '../../../util/util';
export var DEFAULT_INPUTS_O_CARD_RENDERER = [
    'title',
    'subtitle',
    'image',
    'showImage: show-image',
    'action1Text: action-1-text',
    'action2Text: action-2-text'
];
export var DEFAULT_OUTPUTS_O_CARD_RENDERER = [
    'onAction1Click: action-1',
    'onAction2Click: action-2'
];
var OListItemCardRenderer = (function () {
    function OListItemCardRenderer(elRef, _renderer, _injector, _listItem) {
        this.elRef = elRef;
        this._renderer = _renderer;
        this._injector = _injector;
        this._listItem = _listItem;
        this._showImage = true;
        this.onAction1Click = new EventEmitter();
        this.onAction2Click = new EventEmitter();
    }
    OListItemCardRenderer.prototype.modifyMatListItemElement = function () {
        if (this.elRef.nativeElement && this.elRef.nativeElement.parentElement) {
            var matListItem = this.elRef.nativeElement.parentElement.parentElement;
            matListItem.querySelector('.mat-list-text').remove();
            matListItem.classList.add('o-card-item');
        }
    };
    OListItemCardRenderer.prototype.onAction1ButtonClick = function (e) {
        if (Util.isDefined(e)) {
            e.stopPropagation();
        }
        this.onAction1Click.emit(event);
    };
    OListItemCardRenderer.prototype.onAction2ButtonClick = function (e) {
        if (Util.isDefined(e)) {
            e.stopPropagation();
        }
        this.onAction2Click.emit(event);
    };
    OListItemCardRenderer.prototype.compareListHeight = function (height) {
        return (height === this._listItem._list.rowHeight) || undefined;
    };
    Object.defineProperty(OListItemCardRenderer.prototype, "title", {
        get: function () {
            return this._title;
        },
        set: function (val) {
            this._title = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OListItemCardRenderer.prototype, "subtitle", {
        get: function () {
            return this._subtitle;
        },
        set: function (val) {
            this._subtitle = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OListItemCardRenderer.prototype, "image", {
        get: function () {
            return this._image;
        },
        set: function (val) {
            this._image = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OListItemCardRenderer.prototype, "showImage", {
        get: function () {
            return this._showImage;
        },
        set: function (val) {
            this._showImage = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OListItemCardRenderer.prototype, "action1Text", {
        get: function () {
            return this._action1Text;
        },
        set: function (val) {
            this._action1Text = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OListItemCardRenderer.prototype, "action2Text", {
        get: function () {
            return this._action2Text;
        },
        set: function (val) {
            this._action2Text = val;
        },
        enumerable: true,
        configurable: true
    });
    OListItemCardRenderer.DEFAULT_INPUTS_O_CARD_RENDERER = DEFAULT_INPUTS_O_CARD_RENDERER;
    OListItemCardRenderer.DEFAULT_OUTPUTS_O_CARD_RENDERER = DEFAULT_OUTPUTS_O_CARD_RENDERER;
    return OListItemCardRenderer;
}());
export { OListItemCardRenderer };
