// ontimize-web-ngx/ontimize/components/list/renderers/o-list-item-text.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Inject, Injector, forwardRef, ViewEncapsulation, ElementRef, Renderer2, Optional, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OListItemComponent } from '../list-item/o-list-item.component';
import { OListItemTextRenderer } from './o-list-item-text-renderer.class';
export var DEFAULT_INPUTS_O_LIST_ITEM_TEXT = OListItemTextRenderer.DEFAULT_INPUTS_O_TEXT_RENDERER.concat([
    'iconPosition : icon-position'
]);
export var DEFAULT_OUTPUTS_O_LIST_ITEM_TEXT = OListItemTextRenderer.DEFAULT_OUTPUTS_O_TEXT_RENDERER.slice();
var OListItemTextComponent = (function (_super) {
    __extends(OListItemTextComponent, _super);
    function OListItemTextComponent(elRef, _renderer, _injector, _listItem) {
        var _this = _super.call(this, elRef, _renderer, _injector, _listItem) || this;
        _this._listItem = _listItem;
        _this.ICON_POSITION_LEFT = 'left';
        _this.ICON_POSITION_RIGHT = 'right';
        _this.elRef.nativeElement.classList.add('o-list-item-text');
        return _this;
    }
    OListItemTextComponent.prototype.ngOnInit = function () {
        if (!this.iconPosition || [this.ICON_POSITION_LEFT, this.ICON_POSITION_RIGHT].indexOf(this.iconPosition.toLowerCase()) === -1) {
            this.iconPosition = this.ICON_POSITION_RIGHT;
        }
    };
    OListItemTextComponent.prototype.ngAfterViewInit = function () {
        this.modifyMatListItemElement();
    };
    Object.defineProperty(OListItemTextComponent.prototype, "iconPosition", {
        get: function () {
            return this._iconPosition;
        },
        set: function (val) {
            this._iconPosition = val;
        },
        enumerable: true,
        configurable: true
    });
    OListItemTextComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-list-item-text',
                    template: "\n    <div fxLayout=\"row\" fxLayoutAlign=\"center center\">\n      <mat-icon *ngIf=\"icon !== undefined && iconPosition === ICON_POSITION_LEFT\" class=\"material-icons o-list-item-icon\" (click)=\"onActionIconClick($event)\">{{\n        icon }}</mat-icon>\n      <div class=\"mat-list-text\">\n        <h3 matLine class=\"primary-text\"> {{ title }} </h3>\n        <h4 *ngIf=\"primaryText !== undefined\" matLine class=\"primary-text\">{{ primaryText }}</h4>\n        <p *ngIf=\"secondaryText !== undefined\" matLine class=\"secondary-text\">{{ secondaryText }}</p>\n      </div>\n      <mat-icon *ngIf=\"icon !== undefined && iconPosition === ICON_POSITION_RIGHT\" class=\"material-icons o-list-item-icon\" (click)=\"onActionIconClick($event)\">{{\n        icon }}\n      </mat-icon>\n    </div>\n  ",
                    styles: ["\n    .mat-list.selectable[dense] .mat-list-item-content{padding-right:32px}.mat-list.selectable:not([dense]) .mat-list-item-content{padding-right:36px}\n  "],
                    inputs: DEFAULT_INPUTS_O_LIST_ITEM_TEXT,
                    outputs: DEFAULT_OUTPUTS_O_LIST_ITEM_TEXT,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-custom-list-item]': 'true'
                    }
                },] },
    ];
    OListItemTextComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector },
        { type: OListItemComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OListItemComponent; }),] }] }
    ]; };
    return OListItemTextComponent;
}(OListItemTextRenderer));
export { OListItemTextComponent };
var OListItemTextModule = (function () {
    function OListItemTextModule() {
    }
    OListItemTextModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OListItemTextComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OListItemTextComponent]
                },] },
    ];
    return OListItemTextModule;
}());
export { OListItemTextModule };
