// ontimize-web-ngx/ontimize/components/list/renderers/o-list-item-avatar.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ElementRef, forwardRef, Inject, Injector, NgModule, Optional, Renderer2, ViewEncapsulation } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DomSanitizer } from '@angular/platform-browser';
import { OSharedModule } from '../../../shared';
import { OListItemComponent } from '../list-item/o-list-item.component';
import { OListItemTextRenderer } from './o-list-item-text-renderer.class';
export var DEFAULT_INPUTS_O_LIST_ITEM_AVATAR = OListItemTextRenderer.DEFAULT_INPUTS_O_TEXT_RENDERER.concat([
    'avatar',
    'emptyAvatar: empty-avatar',
    'avatarType: avatar-type'
]);
export var DEFAULT_OUTPUTS_O_LIST_ITEM_AVATAR = OListItemTextRenderer.DEFAULT_OUTPUTS_O_TEXT_RENDERER.slice();
var OListItemAvatarComponent = (function (_super) {
    __extends(OListItemAvatarComponent, _super);
    function OListItemAvatarComponent(elRef, _renderer, _injector, _listItem, sanitizer) {
        var _this = _super.call(this, elRef, _renderer, _injector, _listItem) || this;
        _this._listItem = _listItem;
        _this.sanitizer = sanitizer;
        return _this;
    }
    OListItemAvatarComponent.prototype.ngAfterViewInit = function () {
        this.modifyMatListItemElement();
    };
    OListItemAvatarComponent.prototype.ngOnInit = function () {
        var avatarValue = this.avatar;
        if (!this.avatar) {
            avatarValue = this.emptyAvatar;
        }
        else {
            switch (this.avatarType) {
                case 'base64':
                    avatarValue = ('data:image/png;base64,' + ((typeof (avatarValue.bytes) !== 'undefined') ? avatarValue.bytes : avatarValue));
                    break;
                case 'url':
                default:
                    avatarValue = this.avatar;
                    break;
            }
        }
        this.avatarSrc = this.sanitizer.bypassSecurityTrustResourceUrl(avatarValue);
    };
    Object.defineProperty(OListItemAvatarComponent.prototype, "avatarSrc", {
        get: function () {
            return this._avatarSrc;
        },
        set: function (val) {
            this._avatarSrc = val;
        },
        enumerable: true,
        configurable: true
    });
    OListItemAvatarComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-list-item-avatar',
                    template: "\n    <div fxLayout=\"row\" fxLayoutAlign=\"center center\">\n      <img matListAvatar class=\"avatar\" [src]=\"avatarSrc\" [alt]=\"title\">\n      <div class=\"mat-list-text\">\n        <h3 matLine class=\"primary-text\"> {{ title }} </h3>\n        <h4 *ngIf=\"primaryText !== undefined\" matLine class=\"primary-text\">{{ primaryText }}</h4>\n        <p *ngIf=\"secondaryText !== undefined\" matLine class=\"secondary-text\">{{ secondaryText }}</p>\n      </div>\n      <mat-icon *ngIf=\"icon !== undefined\" class=\"material-icons o-list-item-icon\" (click)=\"onActionIconClick($event)\">{{ icon\n        }}\n      </mat-icon>\n    </div>\n  ",
                    styles: ["\n    .o-list-row-action+.o-list-item-avatar{padding:0px 0px 0px 10px}.mat-list.selectable .o-list-item-avatar .mat-list-text{padding-right:0px}.mat-list .mat-list-item .mat-list-item-content .o-list-item-avatar.o-custom-list-item .mat-list-text,.mat-nav-list .mat-list-item .mat-list-item-content .o-list-item-avatar.o-custom-list-item .mat-list-text{padding:0 16px}\n  "],
                    inputs: DEFAULT_INPUTS_O_LIST_ITEM_AVATAR,
                    outputs: DEFAULT_OUTPUTS_O_LIST_ITEM_AVATAR,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-custom-list-item]': 'true',
                        '[class.o-list-item-avatar]': 'true'
                    }
                },] },
    ];
    OListItemAvatarComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector },
        { type: OListItemComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OListItemComponent; }),] }] },
        { type: DomSanitizer }
    ]; };
    return OListItemAvatarComponent;
}(OListItemTextRenderer));
export { OListItemAvatarComponent };
var OListItemAvatarModule = (function () {
    function OListItemAvatarModule() {
    }
    OListItemAvatarModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OListItemAvatarComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OListItemAvatarComponent]
                },] },
    ];
    return OListItemAvatarModule;
}());
export { OListItemAvatarModule };
