// ontimize-web-ngx/ontimize/shared/material/o-mat-error/o-mat-error.js
import { Component, Input, InjectionToken, ElementRef, ChangeDetectorRef, Optional, Inject, ChangeDetectionStrategy, Injector, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatFormFieldModule, MatTooltipModule } from '@angular/material';
import { Codes } from '../../../util/codes';
export var O_MAT_ERROR_OPTIONS = new InjectionToken('o-mat-error-options');
var nextUniqueId = 0;
var OMatErrorComponent = (function () {
    function OMatErrorComponent(injector, elementRef, cd, errorOptions) {
        this.injector = injector;
        this.elementRef = elementRef;
        this.cd = cd;
        this.id = "mat-error-" + nextUniqueId++;
        this.text = '';
        this.errorOptions = errorOptions ? errorOptions : {};
        this.errorType = this.errorOptions.type || 'standard';
    }
    Object.defineProperty(OMatErrorComponent.prototype, "isStandardError", {
        get: function () {
            return this.errorType === Codes.O_MAT_ERROR_STANDARD;
        },
        enumerable: true,
        configurable: true
    });
    OMatErrorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mat-error',
                    template: "\n    <ng-container *ngIf=\"isStandardError\">\n      {{ text }}\n    </ng-container>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        'class': 'mat-error',
                        'role': 'alert',
                        '[attr.id]': 'id'
                    }
                },] },
    ];
    OMatErrorComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ElementRef },
        { type: ChangeDetectorRef },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [O_MAT_ERROR_OPTIONS,] }] }
    ]; };
    OMatErrorComponent.propDecorators = {
        id: [{ type: Input }],
        text: [{ type: Input }]
    };
    return OMatErrorComponent;
}());
export { OMatErrorComponent };
var OMatErrorModule = (function () {
    function OMatErrorModule() {
    }
    OMatErrorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OMatErrorComponent],
                    imports: [MatTooltipModule, MatFormFieldModule, CommonModule],
                    exports: [OMatErrorComponent]
                },] },
    ];
    return OMatErrorModule;
}());
export { OMatErrorModule };
