// ontimize-web-ngx/ontimize/components/bar-menu/menu-group/o-bar-menu-group.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Injector, ElementRef, NgModule, ViewEncapsulation, Inject, forwardRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OBarMenuComponent } from '../o-bar-menu.component';
import { OBaseMenuItemClass } from '../o-base-menu-item.class';
export var DEFAULT_INPUTS_O_BAR_MENU_GROUP = OBaseMenuItemClass.DEFAULT_INPUTS_O_BASE_MENU_ITEM.slice();
var OBarMenuGroupComponent = (function (_super) {
    __extends(OBarMenuGroupComponent, _super);
    function OBarMenuGroupComponent(menu, elRef, injector) {
        var _this = _super.call(this, menu, elRef, injector) || this;
        _this.menu = menu;
        _this.elRef = elRef;
        _this.injector = injector;
        _this.id = 'm_' + String((new Date()).getTime() + Math.random());
        return _this;
    }
    OBarMenuGroupComponent.DEFAULT_INPUTS_O_BAR_MENU_GROUP = DEFAULT_INPUTS_O_BAR_MENU_GROUP;
    OBarMenuGroupComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-bar-menu-group',
                    template: "\n    <div class=\"fake-li\" [class.fake-li-hover]=\"isHovered\">\n      <label [attr.for]=\"id\" class=\"toggle mat-list-item\">\n        <mat-icon class=\"o-bar-menu-group-icon mat-24\">{{ icon }}</mat-icon>\n        {{ title | oTranslate }}\n        <mat-icon class=\"o-bar-menu-group-icon-arrow\" *ngIf=\"!checkbox.checked\">chevron_right</mat-icon>\n        <mat-icon class=\"o-bar-menu-group-icon-arrow\" *ngIf=\"checkbox.checked\">expand_more</mat-icon>\n      </label>\n      <a class=\"o-bar-menu-group-title\">\n        <mat-icon class=\"o-bar-menu-group-icon mat-24\">{{ icon }}</mat-icon> {{ title | oTranslate }}\n        <mat-icon class=\"o-bar-menu-group-icon-arrow\">chevron_right</mat-icon>\n      </a>\n      <input [attr.id]=\"id\" type=\"checkbox\" #checkbox />\n      <div *ngIf=\"!disabled\" class=\"fake-ul mat-primary mat-elevation-z4\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n  ",
                    styles: ["\n    .o-bar-menu-group[disabled=true] .fake-li{opacity:.5}.o-bar-menu-group[disabled=true] .fake-li a.o-bar-menu-group-title{cursor:default}.o-bar-menu-group .fake-li .o-bar-menu-group-icon,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon{display:inline-block;margin:0 10px 0 0}.o-bar-menu-group .fake-li .o-bar-menu-group-icon ~ .o-bar-menu-group-title,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon ~ .o-bar-menu-group-title{padding-left:0}.o-bar-menu-group .fake-li .o-bar-menu-group-icon-arrow,.o-bar-menu-group .fake-ul .o-bar-menu-group-icon-arrow{margin-left:auto}\n  "],
                    inputs: DEFAULT_INPUTS_O_BAR_MENU_GROUP,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-bar-menu-group]': 'true',
                        '[attr.disabled]': 'disabled'
                    }
                },] },
    ];
    OBarMenuGroupComponent.ctorParameters = function () { return [
        { type: OBarMenuComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OBarMenuComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return OBarMenuGroupComponent;
}(OBaseMenuItemClass));
export { OBarMenuGroupComponent };
var OBarMenuGroupModule = (function () {
    function OBarMenuGroupModule() {
    }
    OBarMenuGroupModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OBarMenuGroupComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OBarMenuGroupComponent]
                },] },
    ];
    return OBarMenuGroupModule;
}());
export { OBarMenuGroupModule };
