// ontimize-web-ngx/ontimize/components/input/checkbox/o-checkbox.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { CommonModule } from '@angular/common';
import { Component, ElementRef, forwardRef, Inject, Injector, NgModule, Optional, ViewEncapsulation } from '@angular/core';
import { OSharedModule } from '../../../shared';
import { Util } from '../../../util/util';
import { OFormComponent } from '../../form/o-form.component';
import { OFormValue } from '../../form/OFormValue';
import { DEFAULT_INPUTS_O_FORM_DATA_COMPONENT, DEFAULT_OUTPUTS_O_FORM_DATA_COMPONENT, OFormDataComponent } from '../../o-form-data-component.class';
export var DEFAULT_INPUTS_O_CHECKBOX = [
    'trueValue: true-value',
    'falseValue: false-value',
    'booleanType: boolean-type',
    'color',
    'labelPosition: label-position'
].concat(DEFAULT_INPUTS_O_FORM_DATA_COMPONENT);
export var DEFAULT_OUTPUTS_O_CHECKBOX = DEFAULT_OUTPUTS_O_FORM_DATA_COMPONENT.slice();
var OCheckboxComponent = (function (_super) {
    __extends(OCheckboxComponent, _super);
    function OCheckboxComponent(form, elRef, injector) {
        var _this = _super.call(this, form, elRef, injector) || this;
        _this.trueValue = true;
        _this.falseValue = false;
        _this.booleanType = 'boolean';
        _this.labelPosition = 'after';
        _this._defaultSQLTypeKey = 'BOOLEAN';
        return _this;
    }
    OCheckboxComponent.prototype.initialize = function () {
        _super.prototype.initialize.call(this);
        if (!Util.isDefined(this.sqlType)) {
            switch (this.booleanType) {
                case 'number':
                    this.sqlType = 'INTEGER';
                    break;
                case 'string':
                    this.sqlType = 'VARCHAR';
                    break;
                case 'boolean':
                default:
                    this.sqlType = 'BOOLEAN';
            }
        }
        var context = this;
        this.getFormControl().getValue.bind(context);
        this.getFormControl().getValue = function () {
            return this.value ? context.trueValue : context.falseValue;
        };
    };
    OCheckboxComponent.prototype.ensureOFormValue = function (value) {
        this.parseInputs();
        if (value instanceof OFormValue) {
            if (value.value === undefined) {
                value.value = false;
            }
            this.value = new OFormValue(this.parseValueByType(value.value) === this.trueValue);
        }
        else if (typeof value === 'boolean') {
            this.value = new OFormValue(value);
        }
        else {
            this.value = new OFormValue(this.parseValueByType(value) === this.trueValue);
        }
    };
    OCheckboxComponent.prototype.onClickBlocker = function (evt) {
        evt.stopPropagation();
    };
    OCheckboxComponent.prototype.parseValueByType = function (value) {
        var result;
        switch (this.booleanType) {
            case 'string':
                result = value + '';
                break;
            case 'number':
                result = parseInt(value);
                break;
            default:
                result = value;
                break;
        }
        return result;
    };
    OCheckboxComponent.prototype.parseStringInputs = function () {
        if ((this.trueValue || '').length === 0) {
            this.trueValue = undefined;
        }
        if ((this.falseValue || '').length === 0) {
            this.falseValue = undefined;
        }
    };
    OCheckboxComponent.prototype.parseNumberInputs = function () {
        this.trueValue = parseInt(this.trueValue);
        if (isNaN(this.trueValue)) {
            this.trueValue = 1;
        }
        this.falseValue = parseInt(this.falseValue);
        if (isNaN(this.falseValue)) {
            this.falseValue = 0;
        }
    };
    OCheckboxComponent.prototype.parseInputs = function () {
        switch (this.booleanType) {
            case 'string':
                this.parseStringInputs();
                break;
            case 'number':
                this.parseNumberInputs();
                break;
            default:
                this.trueValue = true;
                this.falseValue = false;
                break;
        }
    };
    OCheckboxComponent.DEFAULT_INPUTS_O_CHECKBOX = DEFAULT_INPUTS_O_CHECKBOX;
    OCheckboxComponent.DEFAULT_OUTPUTS_O_CHECKBOX = DEFAULT_OUTPUTS_O_CHECKBOX;
    OCheckboxComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-checkbox',
                    inputs: DEFAULT_INPUTS_O_CHECKBOX,
                    outputs: DEFAULT_OUTPUTS_O_CHECKBOX,
                    template: "\n    <div [class.custom-width]=\"hasCustomWidth\" [formGroup]=\"getFormGroup()\" class=\"relative\" [matTooltip]=\"tooltip\"\n      [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\" [matTooltipShowDelay]=\"tooltipShowDelay\"\n      [matTooltipHideDelay]=\"tooltipHideDelay\" fxLayout=\"row\" fxLayoutAlign=\"start center\" fxFill>\n      <mat-checkbox [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\" (change)=\"onChangeEvent($event)\"\n        [labelPosition]=\"labelPosition\" [color]=\"color\">\n        {{ olabel | oTranslate }}\n        <mat-error *ngIf=\"hasError('required')\" text=\"{{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\"></mat-error>\n        <mat-error *ngFor=\"let oError of getActiveOErrors()\" text=\"{{ oError.text | oTranslate }}\"></mat-error>\n      </mat-checkbox>\n      <div *ngIf=\"isReadOnly\" (click)=\"onClickBlocker($event)\" class=\"read-only-blocker\" fxFill></div>\n    </div>\n  ",
                    styles: ["\n    .o-checkbox mat-checkbox{z-index:1}.o-checkbox .mat-checkbox-disabled .mat-checkbox-layout .mat-checkbox-label{color:rgba(0,0,0,0.36)}.o-checkbox .custom-width{white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.o-checkbox .read-only-blocker{z-index:2;position:absolute;top:0;left:0;right:0}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-checkbox]': 'true'
                    }
                },] },
    ];
    OCheckboxComponent.ctorParameters = function () { return [
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return OCheckboxComponent;
}(OFormDataComponent));
export { OCheckboxComponent };
var OCheckboxModule = (function () {
    function OCheckboxModule() {
    }
    OCheckboxModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OCheckboxComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OCheckboxComponent]
                },] },
    ];
    return OCheckboxModule;
}());
export { OCheckboxModule };
