// ontimize-web-ngx/ontimize/layouts/app-layout/o-app-layout.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { CommonModule } from '@angular/common';
import { Component, EventEmitter, NgModule, ViewChild, ViewEncapsulation } from '@angular/core';
import { RouterModule } from '@angular/router';
import { OAppHeaderModule } from '../../components/app-header/o-app-header.component';
import { OAppSidenavComponent, OAppSidenavModule } from '../../components/app-sidenav/o-app-sidenav.component';
import { InputConverter } from '../../decorators';
import { OSharedModule } from '../../shared';
import { Util } from '../../util/util';
import { OAppLayoutHeaderComponent } from './app-layout-header/o-app-layout-header.component';
import { OAppLayoutSidenavComponent } from './app-layout-sidenav/o-app-layout-sidenav.component';
export var DEFAULT_INPUTS_O_APP_LAYOUT = [
    'mode',
    'sidenavMode: sidenav-mode',
    'sidenavOpened: sidenav-opened',
    '_showHeader: show-header',
    'showUserInfo: show-user-info',
    'showLanguageSelector: show-language-selector',
    'useFlagIcons: use-flag-icons',
    'openedSidenavImg: opened-sidenav-image',
    'closedSidenavImg: closed-sidenav-image'
];
export var DEFAULT_OUTPUTS_O_APP_LAYOUT = [
    'beforeOpenSidenav',
    'afterOpenSidenav',
    'beforeCloseSidenav',
    'afterCloseSidenav'
];
var OAppLayoutComponent = (function () {
    function OAppLayoutComponent() {
        this.sidenavOpened = true;
        this.showUserInfo = true;
        this.showLanguageSelector = true;
        this.useFlagIcons = false;
        this.beforeOpenSidenav = new EventEmitter();
        this.afterOpenSidenav = new EventEmitter();
        this.beforeCloseSidenav = new EventEmitter();
        this.afterCloseSidenav = new EventEmitter();
    }
    Object.defineProperty(OAppLayoutComponent.prototype, "showHeader", {
        get: function () {
            return this._showHeader;
        },
        set: function (val) {
            this._showHeader = val;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OAppLayoutComponent.prototype, "mode", {
        get: function () {
            return this._mode;
        },
        set: function (val) {
            var m = OAppLayoutComponent.OAppLayoutModes.find(function (e) { return e === val; });
            if (Util.isDefined(m)) {
                this._mode = m;
                if (this._mode === 'mobile' && !Util.isDefined(this.showHeader)) {
                    this.showHeader = true;
                }
            }
            else {
                console.error('Invalid `o-app-layout` mode (' + val + ')');
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OAppLayoutComponent.prototype, "sidenavMode", {
        get: function () {
            return this._sidenavMode;
        },
        set: function (val) {
            var m = OAppLayoutComponent.OSidenavModes.find(function (e) { return e === val; });
            if (Util.isDefined(m)) {
                this._sidenavMode = m;
            }
            else {
                console.error('Invalid `o-app-layout` sidenav-mode (' + val + ')');
            }
        },
        enumerable: true,
        configurable: true
    });
    OAppLayoutComponent.prototype.sidenavToggle = function (opened) {
        opened ? this.beforeOpenSidenav.emit() : this.beforeCloseSidenav.emit();
    };
    OAppLayoutComponent.prototype.afterToggle = function (opened) {
        opened ? this.afterOpenSidenav.emit() : this.afterCloseSidenav.emit();
    };
    OAppLayoutComponent.DEFAULT_INPUTS_O_APP_LAYOUT = DEFAULT_INPUTS_O_APP_LAYOUT;
    OAppLayoutComponent.DEFAULT_OUTPUTS_O_APP_LAYOUT = DEFAULT_OUTPUTS_O_APP_LAYOUT;
    OAppLayoutComponent.OAppLayoutModes = ['mobile', 'desktop'];
    OAppLayoutComponent.OSidenavModes = ['over', 'push', 'side'];
    OAppLayoutComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-app-layout',
                    inputs: DEFAULT_INPUTS_O_APP_LAYOUT,
                    outputs: DEFAULT_OUTPUTS_O_APP_LAYOUT,
                    template: "\n    <o-app-sidenav #appSidenav [sidenav-mode]=\"sidenavMode\" [opened]=\"sidenavOpened\" [show-user-info]=\"showUserInfo\"\n      [show-toggle-button]=\"!showHeader\" [opened-sidenav-image]=\"openedSidenavImg\" [closed-sidenav-image]=\"closedSidenavImg\"\n      [layout-mode]=\"mode\" [class.header-layout]=\"showHeader\" (onSidenavToggle)=\"sidenavToggle($event)\"\n      (afterSidenavToggle)=\"afterToggle($event)\">\n\n      <ng-content select=\"o-app-layout-sidenav[position=start]\" ngProjectAs=\"o-app-layout-sidenav-projection-start\">\n      </ng-content>\n      <ng-content select=\"o-app-layout-sidenav[position=end]\" ngProjectAs=\"o-app-layout-sidenav-projection-end\">\n      </ng-content>\n      <ng-content select=\"o-app-layout-sidenav\" ngProjectAs=\"o-app-layout-sidenav-projection-end\"></ng-content>\n\n      <o-app-header #appHeader *ngIf=\"showHeader\" [show-user-info]=\"showUserInfo\"\n        [show-language-selector]=\"showLanguageSelector\" [use-flag-icons]=\"useFlagIcons\"\n        (onSidenavToggle)=\"appSidenav.toggleSidenav()\">\n\n        <ng-content select=\"o-app-layout-header\"></ng-content>\n\n      </o-app-header>\n\n      <div class=\"application-layout-content-wrapper\" [class.header-layout]=\"showHeader\">\n        <ng-content></ng-content>\n      </div>\n    </o-app-sidenav>\n  ",
                    styles: ["\n    .application-layout-content-wrapper{z-index:1;position:absolute;top:0px;bottom:0px;left:0px;right:0px;overflow:auto;padding:8px}.application-layout-content-wrapper.header-layout{top:56px}\n  "],
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    OAppLayoutComponent.propDecorators = {
        appSidenav: [{ type: ViewChild, args: ['appSidenav',] }]
    };
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "sidenavOpened", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showUserInfo", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "showLanguageSelector", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "useFlagIcons", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppLayoutComponent.prototype, "_showHeader", void 0);
    return OAppLayoutComponent;
}());
export { OAppLayoutComponent };
var OAppLayoutModule = (function () {
    function OAppLayoutModule() {
    }
    OAppLayoutModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, OSharedModule, RouterModule, OAppSidenavModule, OAppHeaderModule],
                    declarations: [OAppLayoutComponent, OAppLayoutHeaderComponent, OAppLayoutSidenavComponent],
                    exports: [OAppLayoutComponent, OAppLayoutHeaderComponent, OAppLayoutSidenavComponent]
                },] },
    ];
    return OAppLayoutModule;
}());
export { OAppLayoutModule };
