// ontimize-web-ngx/ontimize/components/app-sidenav/image/o-app-sidenav-image.component.js
import { Injector, NgModule, Component, ViewEncapsulation, ChangeDetectorRef, ChangeDetectionStrategy } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OAppSidenavComponent } from '../o-app-sidenav.component';
export var DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE = [
    'openedSrc: opened-src',
    'closedSrc: closed-src'
];
export var DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE = [];
var OAppSidenavImageComponent = (function () {
    function OAppSidenavImageComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.sidenav = this.injector.get(OAppSidenavComponent);
    }
    OAppSidenavImageComponent.prototype.ngOnInit = function () {
        if (this.sidenav) {
            var self_1 = this;
            this.sidenavOpenSubs = this.sidenav.sidenav.openedStart.subscribe(function (opened) {
                self_1.updateImage();
            });
            this.sidenavCloseSubs = this.sidenav.sidenav.closedStart.subscribe(function (opened) {
                self_1.updateImage();
            });
        }
        this.updateImage();
    };
    OAppSidenavImageComponent.prototype.ngOnDestroy = function () {
        if (this.sidenavOpenSubs) {
            this.sidenavOpenSubs.unsubscribe();
        }
        if (this.sidenavCloseSubs) {
            this.sidenavCloseSubs.unsubscribe();
        }
    };
    OAppSidenavImageComponent.prototype.updateImage = function () {
        if (this.sidenav && this.sidenav.sidenav.opened) {
            this.setOpenedImg();
        }
        else {
            this.setClosedImg();
        }
        this.cd.detectChanges();
    };
    Object.defineProperty(OAppSidenavImageComponent.prototype, "src", {
        get: function () {
            return this._src;
        },
        set: function (val) {
            this._src = val;
        },
        enumerable: true,
        configurable: true
    });
    OAppSidenavImageComponent.prototype.setOpenedImg = function () {
        this.src = this.openedSrc;
    };
    OAppSidenavImageComponent.prototype.setClosedImg = function () {
        this.src = this.closedSrc;
    };
    Object.defineProperty(OAppSidenavImageComponent.prototype, "showImage", {
        get: function () {
            return (this._src !== undefined && this._src.length > 0);
        },
        enumerable: true,
        configurable: true
    });
    OAppSidenavImageComponent.DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE = DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE;
    OAppSidenavImageComponent.DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE = DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE;
    OAppSidenavImageComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-app-sidenav-image',
                    inputs: DEFAULT_INPUTS_O_APP_SIDENAV_IMAGE,
                    outputs: DEFAULT_OUTPUTS_O_APP_SIDENAV_IMAGE,
                    template: "\n    <div class=\"o-app-sidenav-image-container\" *ngIf=\"showImage\">\n      <img class=\"o-app-sidenav-image\" [src]=\"src\" />\n    </div>\n  ",
                    styles: ["\n    .sidenav-toggle-container+.o-app-sidenav-image .o-app-sidenav-image-container{padding-top:0;text-align:center}.mat-drawer-opened .o-app-sidenav-image .o-app-sidenav-image-container{text-align:center}.o-app-sidenav-image .o-app-sidenav-image-container{max-height:75px;padding:8px 16px 16px}.o-app-sidenav-image .o-app-sidenav-image-container .o-app-sidenav-image{max-width:100%;max-height:100%}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-app-sidenav-image]': 'true'
                    },
                    changeDetection: ChangeDetectionStrategy.OnPush
                },] },
    ];
    OAppSidenavImageComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    return OAppSidenavImageComponent;
}());
export { OAppSidenavImageComponent };
var OAppSidenavImageModule = (function () {
    function OAppSidenavImageModule() {
    }
    OAppSidenavImageModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, OSharedModule],
                    declarations: [OAppSidenavImageComponent],
                    exports: [OAppSidenavImageComponent]
                },] },
    ];
    return OAppSidenavImageModule;
}());
export { OAppSidenavImageModule };
