// ontimize-web-ngx/ontimize/components/bar-menu/locale-menu-item/o-locale-bar-menu-item.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Inject, Injector, forwardRef, ElementRef, NgModule, ViewEncapsulation } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OBarMenuComponent } from '../o-bar-menu.component';
import { OBaseMenuItemClass } from '../o-base-menu-item.class';
export var DEFAULT_INPUTS_O_LOCALE_BAR_MENU_ITEM = OBaseMenuItemClass.DEFAULT_INPUTS_O_BASE_MENU_ITEM.concat([
    'locale'
]);
var OLocaleBarMenuItemComponent = (function (_super) {
    __extends(OLocaleBarMenuItemComponent, _super);
    function OLocaleBarMenuItemComponent(menu, elRef, injector) {
        var _this = _super.call(this, menu, elRef, injector) || this;
        _this.menu = menu;
        _this.elRef = elRef;
        _this.injector = injector;
        return _this;
    }
    OLocaleBarMenuItemComponent.prototype.configureI18n = function () {
        if (this.isConfiguredLang()) {
            return;
        }
        if (this.translateService) {
            this.translateService.use(this.locale);
        }
        if (this.menu) {
            this.menu.collapseAll();
        }
    };
    OLocaleBarMenuItemComponent.prototype.isConfiguredLang = function () {
        if (this.translateService) {
            return (this.translateService.getCurrentLang() === this.locale);
        }
        return false;
    };
    OLocaleBarMenuItemComponent.DEFAULT_INPUTS_O_LOCALE_BAR_MENU_ITEM = DEFAULT_INPUTS_O_LOCALE_BAR_MENU_ITEM;
    OLocaleBarMenuItemComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-locale-bar-menu-item',
                    template: "\n    <div class=\"fake-li mat-primary \" [class.fake-li-hover]=\"isHovered\" [class.is-selected]=\"isConfiguredLang()\">\n      <a fxLayout=\"row\" fxLayoutAlign=\"space-between center\" (click)=\"configureI18n()\" class=\"mat-list-item\">\n        <mat-icon *ngIf=\"icon !== undefined\" class=\"o-bar-menu-item-icon mat-24\">{{ icon }}</mat-icon>\n        <div fxFlex matLine class=\"o-bar-menu-item-title\">{{ title | oTranslate }}</div>\n        <mat-icon *ngIf=\"isConfiguredLang()\" class=\"o-bar-menu-item-icon configured-lang mat-24\" svgIcon=\"ontimize:check_circle\"></mat-icon>\n      </a>\n    </div>\n  ",
                    styles: ["\n    .o-bar-menu-item-icon.configured-lang{flex:0 1 auto;line-height:24px}\n  "],
                    inputs: DEFAULT_INPUTS_O_LOCALE_BAR_MENU_ITEM,
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-locale-bar-menu-item]': 'true'
                    }
                },] },
    ];
    OLocaleBarMenuItemComponent.ctorParameters = function () { return [
        { type: OBarMenuComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OBarMenuComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return OLocaleBarMenuItemComponent;
}(OBaseMenuItemClass));
export { OLocaleBarMenuItemComponent };
var OLocaleBarMenuItemModule = (function () {
    function OLocaleBarMenuItemModule() {
    }
    OLocaleBarMenuItemModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OLocaleBarMenuItemComponent],
                    imports: [OSharedModule, CommonModule, RouterModule],
                    exports: [OLocaleBarMenuItemComponent]
                },] },
    ];
    return OLocaleBarMenuItemModule;
}());
export { OLocaleBarMenuItemModule };
