// ontimize-web-ngx/ontimize/layouts/form-layout/dialog/o-form-layout-dialog.component.js
import { Component, Inject, ViewEncapsulation, ComponentFactoryResolver, ViewChild, Injector } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
import { OFormLayoutManagerContentDirective } from '../directives/o-form-layout-manager-content.directive';
var OFormLayoutDialogComponent = (function () {
    function OFormLayoutDialogComponent(dialogRef, injector, componentFactoryResolver, data) {
        this.dialogRef = dialogRef;
        this.injector = injector;
        this.componentFactoryResolver = componentFactoryResolver;
        if (data.title) {
            this.title = data.title;
        }
        if (data.data) {
            this.data = data.data;
            var component = data.data.component;
            this.componentFactory = this.componentFactoryResolver.resolveComponentFactory(component);
            this.params = data.data.params;
            this.queryParams = data.data.queryParams;
            this.urlSegments = data.data.urlSegments;
        }
        if (data.layoutManagerComponent) {
            this.formLayoutManager = data.layoutManagerComponent;
        }
    }
    OFormLayoutDialogComponent.prototype.ngAfterViewInit = function () {
        if (this.contentDirective && this.componentFactory) {
            var viewContainerRef = this.contentDirective.viewContainerRef;
            viewContainerRef.clear();
            viewContainerRef.createComponent(this.componentFactory);
        }
    };
    OFormLayoutDialogComponent.prototype.updateNavigation = function (data, id) {
        var label = this.formLayoutManager.getLabelFromData(data);
        label = label.length ? label : this.formLayoutManager.getLabelFromUrlParams(this.params);
        if (label && label.length) {
            label = ': ' + label;
        }
        this.label = label;
    };
    OFormLayoutDialogComponent.prototype.updateActiveData = function (data) {
        this.data = Object.assign(this.data, data);
    };
    OFormLayoutDialogComponent.prototype.closeDialog = function () {
        this.dialogRef.close();
    };
    OFormLayoutDialogComponent.prototype.getRouteOfActiveItem = function () {
        var parentRoute = this.formLayoutManager.parentFormLayoutManager.getRouteOfActiveItem();
        var segments = (this.urlSegments || []);
        var route = [];
        segments.forEach(function (segment, index) {
            if (parentRoute[index] !== segment.path) {
                route.push(segment.path);
            }
        });
        return route;
    };
    OFormLayoutDialogComponent.prototype.getParams = function () {
        return this.params;
    };
    OFormLayoutDialogComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-form-layout-dialog',
                    template: "\n    <div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" class=\"title-container\">\n      <span fxFlex fxLayoutAlign=\"center center\" mat-dialog-title> {{ (title || 'LAYOUT_MANANGER.DIALOG_TITLE') | oTranslate }}{{ label }} </span>\n      <mat-icon (click)=\"closeDialog()\" svgIcon=\"ontimize:close\"></mat-icon>\n    </div>\n\n    <div mat-dialog-content class=\"form-layout-dialog-container\">\n      <ng-template o-form-layout-manager-content></ng-template>\n    </div>\n  ",
                    styles: ["\n    .o-form-layout-dialog-overlay{width:65%;height:90%}.o-form-layout-dialog-overlay .title-container{cursor:default;height:64px;padding:0 16px}.o-form-layout-dialog-overlay .title-container .mat-dialog-title{width:100%;padding-left:24px;text-align:center;margin:0}.o-form-layout-dialog-overlay .title-container .mat-icon{cursor:pointer}.o-form-layout-dialog-overlay .mat-dialog-container{padding:0 !important;overflow:hidden}.o-form-layout-dialog-overlay .mat-dialog-container .mat-dialog-content.form-layout-dialog-container{display:block;padding-top:16px;margin:0;max-height:calc(100% - 78px);height:100%}.o-form-layout-dialog-overlay .mat-dialog-container .o-form-layout-dialog o-form-toolbar .mat-toolbar{padding:0}.o-form-layout-dialog-overlay .mat-dialog-container .o-form-layout-dialog .o-form form.inner-form{position:relative;margin-top:0}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-form-layout-dialog]': 'true'
                    }
                },] },
    ];
    OFormLayoutDialogComponent.ctorParameters = function () { return [
        { type: MatDialogRef },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    OFormLayoutDialogComponent.propDecorators = {
        contentDirective: [{ type: ViewChild, args: [OFormLayoutManagerContentDirective,] }]
    };
    return OFormLayoutDialogComponent;
}());
export { OFormLayoutDialogComponent };
