// ontimize-web-ngx/ontimize/components/table/column/cell-renderer/date/o-table-cell-renderer-date.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Injector, ViewChild, TemplateRef, ChangeDetectionStrategy } from '@angular/core';
import { OMomentPipe } from '../../../../../pipes';
import { OBaseTableCellRenderer } from '../o-base-table-cell-renderer.class';
export var DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_DATE = [
    'format'
];
var OTableCellRendererDateComponent = (function (_super) {
    __extends(OTableCellRendererDateComponent, _super);
    function OTableCellRendererDateComponent(injector) {
        var _this = _super.call(this, injector) || this;
        _this.injector = injector;
        _this.tableColumn.type = 'date';
        _this.setComponentPipe();
        return _this;
    }
    OTableCellRendererDateComponent.prototype.setComponentPipe = function () {
        this.componentPipe = new OMomentPipe(this.injector);
    };
    OTableCellRendererDateComponent.prototype.ngOnInit = function () {
        this.pipeArguments = {
            format: this.format
        };
    };
    OTableCellRendererDateComponent.DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_DATE = DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_DATE;
    OTableCellRendererDateComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-table-cell-renderer-date',
                    template: "\n    <ng-template #templateref let-cellvalue=\"cellvalue\">\n       {{getCellData(cellvalue)}}\n    </ng-template>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    inputs: DEFAULT_INPUTS_O_TABLE_CELL_RENDERER_DATE
                },] },
    ];
    OTableCellRendererDateComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    OTableCellRendererDateComponent.propDecorators = {
        templateref: [{ type: ViewChild, args: ['templateref', { read: TemplateRef },] }]
    };
    return OTableCellRendererDateComponent;
}(OBaseTableCellRenderer));
export { OTableCellRendererDateComponent };
