// ontimize-web-ngx/ontimize/components/app-header/o-app-header.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ElementRef, EventEmitter, Injector, NgModule, ViewEncapsulation } from '@angular/core';
import { Router } from '@angular/router';
import { CommonModule } from '@angular/common';
import { ServiceUtils } from '../../utils';
import { OSharedModule } from '../../shared';
import { InputConverter } from '../../decorators';
import { OUserInfoModule } from '../../components';
import { DialogService, OModulesInfoService } from '../../services';
import { OLanguageSelectorModule } from '../language-selector/o-language-selector.component';
export var DEFAULT_INPUTS_O_APP_HEADER = [
    'showUserInfo: show-user-info',
    'showLanguageSelector: show-language-selector',
    'useFlagIcons: use-flag-icons'
];
export var DEFAULT_OUTPUTS_O_APP_HEADER = [
    'onSidenavToggle'
];
var OAppHeaderComponent = (function () {
    function OAppHeaderComponent(router, injector, elRef) {
        var _this = this;
        this.router = router;
        this.injector = injector;
        this.elRef = elRef;
        this._headerTitle = '';
        this.showUserInfo = true;
        this.showLanguageSelector = true;
        this.useFlagIcons = false;
        this.onSidenavToggle = new EventEmitter();
        this.dialogService = this.injector.get(DialogService);
        this.modulesInfoService = this.injector.get(OModulesInfoService);
        this.modulesInfoSubscription = this.modulesInfoService.getModuleChangeObservable().subscribe(function (res) {
            _this.headerTitle = res.name;
        });
    }
    OAppHeaderComponent.prototype.ngOnDestroy = function () {
        this.modulesInfoSubscription.unsubscribe();
    };
    OAppHeaderComponent.prototype.onLogoutClick = function () {
        var _this = this;
        this.dialogService.confirm('CONFIRM', 'MESSAGES.CONFIRM_LOGOUT').then(function (res) {
            if (res) {
                ServiceUtils.redirectLogin(_this.router, false);
            }
        });
    };
    Object.defineProperty(OAppHeaderComponent.prototype, "headerTitle", {
        get: function () {
            return this._headerTitle;
        },
        set: function (value) {
            this._headerTitle = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OAppHeaderComponent.prototype, "showHeaderTitle", {
        get: function () {
            return this._headerTitle.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    OAppHeaderComponent.DEFAULT_INPUTS_O_APP_HEADER = DEFAULT_INPUTS_O_APP_HEADER;
    OAppHeaderComponent.DEFAULT_OUTPUTS_O_APP_HEADER = DEFAULT_OUTPUTS_O_APP_HEADER;
    OAppHeaderComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-app-header',
                    inputs: DEFAULT_INPUTS_O_APP_HEADER,
                    outputs: DEFAULT_OUTPUTS_O_APP_HEADER,
                    template: "\n    <nav fxFlex fxLayout=\"row\" fxLayoutAlign=\"space-between stretch\">\n      <mat-icon class=\"sidenav-toggle\" (click)=\"onSidenavToggle.emit()\" svgIcon=\"ontimize:menu\"></mat-icon>\n      <div fxFlex fxLayout=\"row\" fxLayoutAlign=\"space-between stretch\">\n        <ng-content></ng-content>\n        <span *ngIf=\"showHeaderTitle\" class=\"o-app-header-title\" fxLayout=\"row\" fxLayoutAlign=\"center center\">\n          {{ headerTitle | oTranslate }}\n        </span>\n        <div fxLayout=\"row\" fxLayoutAlign=\"end stretch\" class=\"o-app-header-default-actions\"> \n          <o-user-info *ngIf=\"showUserInfo\"></o-user-info>\n          <o-language-selector *ngIf=\"showLanguageSelector\" [use-flag-icons]=\"useFlagIcons\"></o-language-selector>\n        </div>\n      </div>\n    </nav>\n  ",
                    styles: ["\n    .o-app-header{z-index:2;box-shadow:0 2px 6px rgba(0,0,0,0.24);height:56px}.o-app-header nav{padding-right:16px}.o-app-header nav .sidenav-toggle{text-align:center;padding:16px;cursor:pointer}.o-app-header nav .o-app-header-title{padding:0 16px;cursor:default}.o-app-header nav .o-app-header-default-actions{margin-left:auto}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-app-header]': 'true'
                    }
                },] },
    ];
    OAppHeaderComponent.ctorParameters = function () { return [
        { type: Router },
        { type: Injector },
        { type: ElementRef }
    ]; };
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "showUserInfo", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "showLanguageSelector", void 0);
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OAppHeaderComponent.prototype, "useFlagIcons", void 0);
    return OAppHeaderComponent;
}());
export { OAppHeaderComponent };
var OAppHeaderModule = (function () {
    function OAppHeaderModule() {
    }
    OAppHeaderModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, OLanguageSelectorModule, OUserInfoModule, OSharedModule],
                    declarations: [OAppHeaderComponent],
                    exports: [OAppHeaderComponent]
                },] },
    ];
    return OAppHeaderModule;
}());
export { OAppHeaderModule };
