// ontimize-web-ngx/ontimize/components/button/o-button.component.js
import { CommonModule } from '@angular/common';
import { Component, NgModule, ViewEncapsulation } from '@angular/core';
import { OSharedModule } from '../../shared';
export var DEFAULT_INPUTS_O_BUTTON = [
    'oattr: attr',
    'olabel: label',
    'otype: type',
    'icon',
    'svgIcon : svg-icon',
    'iconPosition: icon-position',
    'image'
];
var OButtonComponent = (function () {
    function OButtonComponent() {
        this.otype = OButtonComponent.DEFAULT_TYPE;
    }
    OButtonComponent.prototype.ngOnInit = function () {
        if (this.otype) {
            this.otype = this.otype.toUpperCase();
        }
    };
    Object.defineProperty(OButtonComponent.prototype, "needsIconButtonClass", {
        get: function () {
            return this.icon !== undefined && (this.olabel === undefined || this.olabel === '');
        },
        enumerable: true,
        configurable: true
    });
    OButtonComponent.prototype.isFab = function () {
        return this.otype === 'FAB';
    };
    OButtonComponent.prototype.isRaised = function () {
        return this.otype === 'RAISED';
    };
    OButtonComponent.prototype.isFlat = function () {
        return this.otype === 'FLAT';
    };
    OButtonComponent.prototype.isStroked = function () {
        return (this.otype === 'STROKED' || !this.otype);
    };
    OButtonComponent.prototype.isBasic = function () {
        return this.otype === 'BASIC';
    };
    OButtonComponent.prototype.isMiniFab = function () {
        return this.otype === 'FAB-MINI';
    };
    OButtonComponent.prototype.isIconButton = function () {
        return this.otype === 'ICON';
    };
    OButtonComponent.DEFAULT_INPUTS_O_BUTTON = DEFAULT_INPUTS_O_BUTTON;
    OButtonComponent.DEFAULT_TYPE = 'STROKED';
    OButtonComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-button',
                    inputs: DEFAULT_INPUTS_O_BUTTON,
                    template: "\n    <button type=\"button\" *ngIf=\"isBasic()\" mat-button [class.mat-icon-button]=\"needsIconButtonClass\">\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n        [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n\n    <button type=\"button\" *ngIf=\"isRaised()\" mat-raised-button class=\"mat-raised-button\"\n      [class.mat-icon-button]=\"needsIconButtonClass\">\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n\n    <button type=\"button\" *ngIf=\"isStroked()\" mat-stroked-button [class.mat-icon-button]=\"needsIconButtonClass\">\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n\n    <button type=\"button\" *ngIf=\"isFlat()\" mat-flat-button [class.mat-icon-button]=\"needsIconButtonClass\">\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n\n    <button type=\"button\" *ngIf=\"isIconButton()\" mat-icon-button [class.mat-icon-button]=\"needsIconButtonClass\">\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n    </button>\n\n    <button type=\"button\" *ngIf=\"isFab()\" mat-fab>\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n\n    <button type=\"button\" *ngIf=\"isMiniFab()\" mat-mini-fab>\n      <img src=\"{{ image }}\" class=\"o-button-image\" *ngIf=\"image!=undefined\" />\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"icon!==undefined\" style=\"vertical-align:middle\">{{ icon }}\n      </mat-icon>\n      <mat-icon class=\"material-icons o-button-icon\" *ngIf=\"svgIcon!==undefined\" style=\"vertical-align:middle\"\n      [svgIcon]=\"svgIcon\"></mat-icon>\n      <br *ngIf=\"iconPosition==='top'\" />\n      {{ olabel | oTranslate }}\n    </button>\n  ",
                    styles: ["\n    .o-button{display:inline-block}.o-button .mat-icon-button .mat-button-ripple.mat-ripple{border-radius:50%}\n  "],
                    encapsulation: ViewEncapsulation.None,
                    host: {
                        '[class.o-button]': 'true'
                    }
                },] },
    ];
    OButtonComponent.ctorParameters = function () { return []; };
    return OButtonComponent;
}());
export { OButtonComponent };
var OButtonModule = (function () {
    function OButtonModule() {
    }
    OButtonModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OButtonComponent],
                    imports: [CommonModule, OSharedModule],
                    exports: [OButtonComponent]
                },] },
    ];
    return OButtonModule;
}());
export { OButtonModule };
