// ontimize-web-ngx/ontimize/components/input/nif-input/o-nif-input.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, Inject, Injector, forwardRef, ElementRef, Optional, NgModule, ViewEncapsulation } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OSharedModule } from '../../../shared';
import { OValidators } from '../../../validators/o-validators';
import { OFormComponent } from '../../form/o-form.component';
import { OTextInputModule, OTextInputComponent, DEFAULT_INPUTS_O_TEXT_INPUT, DEFAULT_OUTPUTS_O_TEXT_INPUT } from '../text-input/o-text-input.component';
export var DEFAULT_INPUTS_O_NIF_INPUT = DEFAULT_INPUTS_O_TEXT_INPUT.slice();
export var DEFAULT_OUTPUTS_O_NIF_INPUT = DEFAULT_OUTPUTS_O_TEXT_INPUT.slice();
var ONIFInputComponent = (function (_super) {
    __extends(ONIFInputComponent, _super);
    function ONIFInputComponent(form, elRef, injector) {
        return _super.call(this, form, elRef, injector) || this;
    }
    ONIFInputComponent.prototype.resolveValidators = function () {
        var validators = _super.prototype.resolveValidators.call(this);
        validators.push(OValidators.nifValidator);
        return validators;
    };
    ONIFInputComponent.DEFAULT_INPUTS_O_NIF_INPUT = DEFAULT_INPUTS_O_NIF_INPUT;
    ONIFInputComponent.DEFAULT_OUTPUTS_O_NIF_INPUT = DEFAULT_OUTPUTS_O_NIF_INPUT;
    ONIFInputComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-nif-input',
                    template: "\n    <div fxLayout=\"row\" fxLayoutAlign=\"space-between center\" [formGroup]=\"getFormGroup()\" [matTooltip]=\"tooltip\"\n      [matTooltipClass]=\"tooltipClass\" [matTooltipPosition]=\"tooltipPosition\" [matTooltipShowDelay]=\"tooltipShowDelay\"\n      [matTooltipHideDelay]=\"tooltipHideDelay\">\n      <mat-form-field [appearance]=\"appearance\" [floatLabel]=\"floatLabel\" [hideRequiredMarker]=\"hideRequiredMarker\"\n        [class.custom-width]=\"hasCustomWidth\" class=\"icon-field\" fxFlexFill>\n        <mat-label *ngIf=\"labelVisible\">{{ olabel | oTranslate }}</mat-label>\n        <input matInput type=\"text\" [id]=\"getAttribute()\" [formControlName]=\"getAttribute()\" [placeholder]=\"placeHolder\"\n          (focus)=\"innerOnFocus($event)\" (blur)=\"innerOnBlur($event)\" (change)=\"onChangeEvent($event)\"\n          [readonly]=\"isReadOnly\" [required]=\"isRequired\">\n        <button type=\"button\" *ngIf=\"showClearButton\" matSuffix mat-icon-button (click)=\"onClickClearValue($event)\">\n          <mat-icon svgIcon=\"ontimize:close\"></mat-icon>\n        </button>\n        <mat-icon matSuffix [class.mat-disabled]=\"!enabled\" svgIcon=\"ontimize:perm_identity\"></mat-icon>\n        <mat-error *ngIf=\"hasError('required')\" text=\"{{ 'FORM_VALIDATION.REQUIRED' | oTranslate }}\"></mat-error>\n        <mat-error *ngIf=\"hasError('invalidNIF')\" text=\"{{ 'FORM_VALIDATION.NIF_FORMAT' | oTranslate}}\"></mat-error>\n        <mat-error *ngIf=\"hasError('invalidNIFLetter')\" text=\"{{ 'FORM_VALIDATION.DNI_LETTER' | oTranslate}}\"></mat-error>\n        <mat-error *ngFor=\"let oError of getActiveOErrors()\" text=\"{{ oError.text | oTranslate }}\"></mat-error>\n      </mat-form-field>\n    </div>\n  ",
                    styles: ["\n\n  "],
                    inputs: DEFAULT_INPUTS_O_NIF_INPUT,
                    outputs: DEFAULT_OUTPUTS_O_NIF_INPUT,
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    ONIFInputComponent.ctorParameters = function () { return [
        { type: OFormComponent, decorators: [{ type: Optional }, { type: Inject, args: [forwardRef(function () { return OFormComponent; }),] }] },
        { type: ElementRef },
        { type: Injector }
    ]; };
    return ONIFInputComponent;
}(OTextInputComponent));
export { ONIFInputComponent };
var ONIFInputModule = (function () {
    function ONIFInputModule() {
    }
    ONIFInputModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ONIFInputComponent],
                    imports: [OSharedModule, CommonModule, OTextInputModule],
                    exports: [ONIFInputComponent, OTextInputModule]
                },] },
    ];
    return ONIFInputModule;
}());
export { ONIFInputModule };
