// ontimize-web-ngx/ontimize/components/bar-menu/menu-nested/o-bar-menu-nested.component.js
import { Component, Injector, ViewEncapsulation, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { AppMenuService } from '../../../services/app-menu.service';
import { OBarMenuGroupModule } from '../menu-group/o-bar-menu-group.component';
import { OBarMenuItemModule } from '../menu-item/o-bar-menu-item.component';
export var DEFAULT_INPUTS_O_BAR_MENU_NESTED = [
    'items'
];
var OBarMenuNestedComponent = (function () {
    function OBarMenuNestedComponent(injector) {
        this.injector = injector;
        this.appMenuService = this.injector.get(AppMenuService);
    }
    OBarMenuNestedComponent.prototype.getValueOfAttr = function (menu, attr) {
        var valAttr = '';
        if (menu.hasOwnProperty(attr)) {
            valAttr = menu[attr];
        }
        return valAttr;
    };
    OBarMenuNestedComponent.prototype.isMenuGroup = function (item) {
        return this.appMenuService.getMenuItemType(item) === 'group';
    };
    OBarMenuNestedComponent.DEFAULT_INPUTS_O_BAR_MENU = DEFAULT_INPUTS_O_BAR_MENU_NESTED;
    OBarMenuNestedComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-bar-menu-nested',
                    template: "\n    <ng-container *ngFor=\"let item of items\">\n      <!--menu-group-->\n      <o-bar-menu-group *ngIf=\"isMenuGroup(item)\" [title]=\"getValueOfAttr(item, 'name')\" [attr]=\"getValueOfAttr(item,'id')\"\n        [tooltip]=\"getValueOfAttr(item, 'tooltip')\" [icon]=\"getValueOfAttr(item,'icon')\" [ngClass]=\"item.class\">\n        <o-bar-menu-nested [items]=\"item.items\"> </o-bar-menu-nested>\n      </o-bar-menu-group>\n      <!--menu-item-->\n      <o-bar-menu-item *ngIf=\"!isMenuGroup(item)\" [title]=\"getValueOfAttr(item,'name')\" [attr]=\"getValueOfAttr(item,'id')\"\n        [tooltip]=\"getValueOfAttr(item, 'tooltip') \" [icon]=\"getValueOfAttr(item, 'icon') \"\n        [route]=\"getValueOfAttr(item, 'route') \" [ngClass]=\"item.class\">\n      </o-bar-menu-item>\n    </ng-container>\n  ",
                    inputs: DEFAULT_INPUTS_O_BAR_MENU_NESTED,
                    encapsulation: ViewEncapsulation.None
                },] },
    ];
    OBarMenuNestedComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OBarMenuNestedComponent;
}());
export { OBarMenuNestedComponent };
var OBarMenuNestedModule = (function () {
    function OBarMenuNestedModule() {
    }
    OBarMenuNestedModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OBarMenuNestedComponent],
                    imports: [CommonModule, OBarMenuGroupModule, OBarMenuItemModule],
                    exports: [OBarMenuNestedComponent]
                },] },
    ];
    return OBarMenuNestedModule;
}());
export { OBarMenuNestedModule };
