// ontimize-web-ngx/ontimize/layouts/form-layout/tabgroup/options/o-form-layout-tabgroup-options.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Component, ContentChild, forwardRef, Inject, TemplateRef, ViewEncapsulation } from '@angular/core';
import { InputConverter } from '../../../../decorators/input-converter';
import { OFormLayoutManagerComponent } from '../../o-form-layout-manager.component';
export var DEFAULT_INPUTS_O_FORM_LAYOUT_TABGROUP_OPTIONS = [
    'backgroundColor:background-color',
    'color',
    'headerPosition:header-position',
    'disableAnimation:disable-animation',
    'icon',
    'iconPosition:icon-position'
];
var OFormLayoutTabGroupOptionsComponent = (function () {
    function OFormLayoutTabGroupOptionsComponent(oFormLayoutManager) {
        this.oFormLayoutManager = oFormLayoutManager;
        this.iconPosition = 'left';
    }
    OFormLayoutTabGroupOptionsComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-form-layout-tabgroup-options',
                    template: ' ',
                    encapsulation: ViewEncapsulation.None,
                    inputs: DEFAULT_INPUTS_O_FORM_LAYOUT_TABGROUP_OPTIONS,
                    host: {
                        '[class.o-form-layout-tabgroup-options]': 'true'
                    }
                },] },
    ];
    OFormLayoutTabGroupOptionsComponent.ctorParameters = function () { return [
        { type: OFormLayoutManagerComponent, decorators: [{ type: Inject, args: [forwardRef(function () { return OFormLayoutManagerComponent; }),] }] }
    ]; };
    OFormLayoutTabGroupOptionsComponent.propDecorators = {
        templateMatTabLabel: [{ type: ContentChild, args: [TemplateRef,] }]
    };
    __decorate([
        InputConverter(),
        __metadata("design:type", Boolean)
    ], OFormLayoutTabGroupOptionsComponent.prototype, "disableAnimation", void 0);
    return OFormLayoutTabGroupOptionsComponent;
}());
export { OFormLayoutTabGroupOptionsComponent };
