// ontimize-web-ngx/ontimize/components/contextmenu/context-menu-group/o-context-menu-group.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { OComponentMenuItems } from '../o-content-menu.class';
import { DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM, OContextMenuItemComponent } from '../o-context-menu-components';
export var DEFAULT_CONTEXT_MENU_GROUP_INPUTS = DEFAULT_INPUTS_O_CONTEXT_MENU_ITEM.concat([
    'children'
]);
var OContextMenuGroupComponent = (function (_super) {
    __extends(OContextMenuGroupComponent, _super);
    function OContextMenuGroupComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = OComponentMenuItems.TYPE_GROUP_MENU;
        _this.children = [];
        return _this;
    }
    OContextMenuGroupComponent.prototype.ngAfterContentInit = function () {
        this.children = this.oContextMenuItems.toArray().slice(1, this.oContextMenuItems.toArray().length);
    };
    OContextMenuGroupComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-context-menu-group',
                    template: ' ',
                    inputs: DEFAULT_CONTEXT_MENU_GROUP_INPUTS,
                    providers: [{ provide: OComponentMenuItems, useExisting: forwardRef(function () { return OContextMenuGroupComponent; }) }]
                },] },
    ];
    OContextMenuGroupComponent.propDecorators = {
        oContextMenuItems: [{ type: ContentChildren, args: [OComponentMenuItems,] }]
    };
    return OContextMenuGroupComponent;
}(OContextMenuItemComponent));
export { OContextMenuGroupComponent };
