// ontimize-web-ngx/ontimize/components/input/validation/o-validator.component.js
import { Component, Injector, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { Util } from '../../../utils';
import { OSharedModule } from '../../../shared';
export var DEFAULT_INPUTS_O_VALIDATOR = [
    'validatorFn: validator-function',
    'errorName: error-name',
    'errorText: error-text'
];
var OValidatorComponent = (function () {
    function OValidatorComponent(injector) {
        this.injector = injector;
        this.validatorFn = null;
        this.errorsData = [];
    }
    OValidatorComponent.prototype.registerError = function (oError) {
        this.errorsData.push({
            name: oError.getName(),
            text: oError.getText()
        });
    };
    OValidatorComponent.prototype.getValidatorFn = function () {
        return this.validatorFn;
    };
    OValidatorComponent.prototype.getErrorsData = function () {
        var result = [];
        if (this.errorsData.length > 0) {
            result = this.errorsData;
        }
        else if (Util.isDefined(this.errorName) && Util.isDefined(this.errorText)) {
            result = [{
                    name: this.errorName,
                    text: this.errorText
                }];
        }
        return result;
    };
    OValidatorComponent.DEFAULT_INPUTS_O_VALIDATOR = DEFAULT_INPUTS_O_VALIDATOR;
    OValidatorComponent.decorators = [
        { type: Component, args: [{
                    moduleId: module.id,
                    selector: 'o-validator',
                    template: ' ',
                    inputs: DEFAULT_INPUTS_O_VALIDATOR
                },] },
    ];
    OValidatorComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    return OValidatorComponent;
}());
export { OValidatorComponent };
var OValidatorModule = (function () {
    function OValidatorModule() {
    }
    OValidatorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [OValidatorComponent],
                    imports: [OSharedModule, CommonModule],
                    exports: [OValidatorComponent]
                },] },
    ];
    return OValidatorModule;
}());
export { OValidatorModule };
