// ontimize-web-ngx/ontimize/components/table/extensions/dialog/o-table-base-dialog.class.js
import { QueryList, ViewChildren } from '@angular/core';
import { Codes, Util } from '../../../../utils';
import { O_MAT_ERROR_OPTIONS, OMatErrorComponent } from '../../../../shared';
var OTableBaseDialogClass = (function () {
    function OTableBaseDialogClass(injector) {
        this.injector = injector;
        try {
            this.errorOptions = this.injector.get(O_MAT_ERROR_OPTIONS) || {};
        }
        catch (e) {
            this.errorOptions = {};
        }
    }
    OTableBaseDialogClass.prototype.setFormControl = function (formControl) {
        this.formControl = formControl;
    };
    Object.defineProperty(OTableBaseDialogClass.prototype, "tooltipClass", {
        get: function () {
            var result;
            var liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
            if (liteError && Util.isDefined(this.formControl) && this.formControlHasErrors()) {
                result = "o-tooltip o-mat-error";
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OTableBaseDialogClass.prototype, "tooltipText", {
        get: function () {
            var result;
            var liteError = this.errorOptions.type === Codes.O_MAT_ERROR_LITE;
            if (liteError && this.formControlHasErrors() && this.oMatErrorChildren && this.oMatErrorChildren.length > 0) {
                result = '';
                this.oMatErrorChildren.forEach(function (oMatError) {
                    result += oMatError.text + "\n";
                });
            }
            return result;
        },
        enumerable: true,
        configurable: true
    });
    OTableBaseDialogClass.prototype.formControlHasErrors = function () {
        return Util.isDefined(this.formControl) && this.formControl.touched && Util.isDefined(this.formControl.errors);
    };
    OTableBaseDialogClass.propDecorators = {
        oMatErrorChildren: [{ type: ViewChildren, args: [OMatErrorComponent,] }]
    };
    return OTableBaseDialogClass;
}());
export { OTableBaseDialogClass };
